\section{Introduction}

\software{ssu-align} identifies and aligns small subunit ribosomal RNA
(SSU rRNA) genes in sequence datasets. It uses models of the primary
sequence and secondary structure conservation of SSU rRNA as inferred
by comparative sequence analysis and confirmed by crystal structure
determination from the \db{comparative rna website} (\db{crw}) 
(\htmladdnormallink{http://www.rna.ccbb.utexas.edu/}{http://www.rna.ccbb.utexas.edu/})
\cite{CannoneGutell02}.

\subsection{Overview}
The \software{ssu-align} package contains six different
programs: \textbf{ssu-align}\footnote{SSU-ALIGN is the name of the
  software package as well as the name of one of the programs within
  the package. In this guide, when written in lowercase bold-faced
  type (\textbf{ssu-align}) it refers to the executable program, and
  when written in all capital letters (\sft{SSU-ALIGN}) it refers to
  the package.},
\textbf{ssu-build}, \textbf{ssu-draw}, 
\textbf{ssu-mask}, \textbf{ssu-merge}, and \textbf{ssu-prep}. 

You can use \textbf{ssu-align} to create multiple alignments of SSU
sequences with profile probabilistic models called covariance models
(CMs). The package includes three SSU rRNA CMs, an archaeal model,
a bacterial model, and a eukaryotic (nuclear) model. These are the
default models used, but others, including truncated versions of the
default models that represent only a specific region of SSU (such as
the V4 hypervariable region), can be created using the
\textbf{ssu-build} program.

Given an input dataset of unaligned SSU sequences, \textbf{ssu-align}
proceeds through two stages to generate alignments as depicted in
figure~\ref{fig:strategy}.  
%In stage 1, each of the sequences is aligned to each of
%the three models based on primary sequence conservation to obtain a
%score for each sequence to each model.  The model that gives the
%highest scoring alignment to each sequence is the \emph{best-matching
%  model} for that sequence.  In stage 2, each sequence is aligned to
%its best-matching model using both primary sequence and secondary
%structure conservation information. The end result is a separate
%multiple alignment for each model that was the best-matching model for
%at least one sequence in the input dataset.
In the first stage, each target sequence is scored with each of the
models based only on sequence conservation. This is done with a
profile HMM derived from each CM, which is significantly faster than
using the CM\@.  The model whose profile HMM gives the highest score to
each sequence is defined as the \emph{best-matching} model for that
sequence. If this highest score is not above a predefined threshold,
the sequence is discarded and not evaluated further. The boundaries of
the best-matching HMM alignment are used to truncate each target
sequence if the alignment does not span the entire target length.  In
stage 2, each surviving, and possibly truncated, sequence is aligned
to its best-matching model, this time using the CM which scores
both sequence and conserved structure. Up to $N$ alignments are
created, one for each model that was the best-match to at least one
target sequence.

\begin{figure}
  \begin{center}
\includegraphics[width=6.5in]{Figures/ssualign-schematic}
  \end{center}
\caption{\textbf{Schematic of the \prog{ssu-align} alignment
    pipeline.} Unaligned target sequences are input to the
  program. In stage 1, each sequence is independently aligned
  using only primary sequence scoring to each of $N$ HMMs (by
  default, $N$ is 3), one built from each model in the input
  CM file. The model whose HMM alignment has the maximum bit
  score is the ``best-matching'' model for that sequence, in
  this example ``eukarya'' is the best-matching model. In
  stage 2, the unaligned (sub)sequence from the best-matching
  model's HMM alignment (potentially with some sequence
  trimmed off the ends) is aligned to the best-matching
  model's CM which scores both sequence and conserved
  secondary structure. The CM aligned target sequence is added
  to that model's output alignment. After all target sequences
  are processed, the program has output up to $N$ new
  structural alignments, one for each model that was the
  best-matching model for at least 1 target sequence.}
\label{fig:strategy}
\end{figure}

The \textbf{ssu-mask} program can then be used to removing unreliable
alignment columns. This step is important prior to using the
alignments as input to phylogenetic inference tools that rely heavily
on alignment accuracy.  Alignment columns are selected for removal
based on automatically calculated ``confidence estimates'' for the
aligned residues, which are the probabilities that each residue
belongs in each column of the alignment given the CM. This is
discussed in more detail in the section 3 of this guide.

The \textbf{ssu-draw} program can be used to create secondary
structure diagrams of the alignments created by 
\textbf{ssu-align}. Diagrams that display per-column alignment
statistics, such as information content and frequency of insertions
and deletions, as well as those that display individual aligned
sequences can be drawn.

The \textbf{ssu-prep} and \textbf{ssu-merge} programs allow users to
divide up large alignment jobs into sets of parallel
\textbf{ssu-align} jobs on clusters or multi-core computers.

\subsection{What distinguishes \sft{ssu-align} from other SSU
  alignment tools}

\sft{ssu-align} is a freely available, open source software program
for creating large-scale alignments of small subunit ribosomal RNA
(SSU) using covariance models (CMs). The archaeal, bacterial,
%chloroplast, 
and eukaryotic (nuclear) 
%and mitochondrial 
SSU CMs in \sft{ssu-align} are built from structural seed alignments derived from
Robin Gutell's Comparative RNA Website (CRW) \cite{CannoneGutell02}.
%Though only a prototype which I hope to develop and extend in the
%future, 
The following features distinguish \sft{ssu-align} from other
SSU alignment programs (reviewed in chapter 7 of \cite{Nawrocki09b}):

\begin{itemize}

\item \textbf{Structural alignments are calculated at roughly one
  second per sequence.}  

  CM alignment takes into account conserved sequence and well-nested
  structure. In the past, the slow speed of CM alignment has prevented
  their application to SSU alignment, but the sequence-based banding
  approach described in chapter 8 of \cite{Nawrocki09b} yields roughly a 1000-fold speedup,
  making large-scale SSU CM alignment feasible.

\item \textbf{Masks for pruning ambiguously aligned columns are
  automatically generated.}
  
  CMs are probabilistic models that allow calculation of the posterior
  probability that each sequence nucleotide belongs in each column of the
  output alignment given the model. Regions of low posterior
  probabilities are indicative of high alignment ambiguity and
  should be pruned away (masked out) prior to phylogenetic
  inference. \sft{ssu-align} can automatically detect and remove
  these columns for any alignment it generates.

\item \textbf{Accurate alignments can be computed using seed
  alignments of less than one hundred sequences.}

  CMs built from small seed alignments %of only dozens of sequences
  can achieve comparable accuracy to nearest-neighbor based alignment
  methods that use seed (reference) alignments of thousands or tens of
  thousands of sequences. This reduces the level of manual curation
  necessary for creating useful seeds, making it easier to extend
  \sft{SSU-align} by adding new seeds and corresponding CMs that cover
  specific phylogenetic ranges. For example, a
  \emph{Firmicute}-specific seed and model could be constructed to
  create large alignments for phylogenetic analysis of the \emph{Firmicute}
  bacterial phyla.

\end{itemize}


\subsection{What is included in this guide}

Section~\ref{sec:install} describes how to install the
package. Section~\ref{sec:background} provides background
information on how \sft{ssu-align} aligns SSU sequences and how it
identifies and removes columns from (masks) those alignments.
The tutorial in section~\ref{sec:tutorial} walks through examples of using
the programs in \sft{ssu-align} to create alignments, 
mask alignments, build models of a specific region of SSU, draw
structure diagrams for alignments, and split up large alignment jobs
into smaller jobs to run in parallel on a cluster.
Section~\ref{sec:models} describes how the three default models were
derived from \sft{crw} and how the default masks for those models were
determined. Section~\ref{sec:output} describes the
format of all the output files \sft{ssu-align} creates. Finally, the
\sft{ssu-align} manual pages included at the end of this guide explain
the command-line options of the six programs.

\subsection{What is included in this package}

\sft{ssu-align} includes the \sft{infernal} software package, which is
written in C and is installed along with \sft{ssu-align} by following
the directions in the Installation section of this guide. The six
\sft{ssu-align} programs: \prog{ssu-align}, \prog{ssu-build},
\prog{ssu-draw}, \prog{ssu-mask}, \prog{ssu-merge} and \prog{ssu-prep}
are \sft{perl} scripts.  Also included are the three default (archaea,
bacteria, and eukarya) SSU models and the \emph{seed} alignments they
were built from. These alignments were based on SSU structures and
alignments from the Comparative RNA Website (CRW)
\cite{CannoneGutell02}. A broader discussion of these models and
alignments and the specific procedure used to create them from the CRW
data is explained in section~\ref{section:chap9}.

Installation of the \sft{perl} (Practical Extraction and Report Language,
Larry Wall) interpreter package version 5.0 or later is required to
run the \software{ssu-align} \sft{perl} scripts.

\subsection{What this package requires on your system}

\sft{ssu-align} was developed and tested on Mac OS/X and Linux
(Redhat), but it should work on most UNIX platforms. It requires
\software{perl} version 5.0 or later. No other programs or libraries
are required. If the program \prog{ps2pdf} is installed on
your system and in your PATH, \prog{ssu-draw} and \prog{ssu-mask} will
use it to convert postscript diagrams to PDF, but it is not required. 

\subsection{What this package does not do}

\textsc{ssu-align} only creates alignments, masks them, and draws SSU
secondary structure diagrams; it does not infer trees from the
alignments it creates, nor does it classify sequences beyond reporting
which model in the input CM file they score highest to.

\begin{comment}
\subsection{Other useful references}

The \software{infernal} user's guide \cite{infernalguide} included in
this package supplements this user's guide. My Ph.D. thesis 
(\htmladdnormallink{http://selab.janelia.org/publications.html}{http://selab.janelia.org/publications.html})
introduces and describes my work on CM methods. It includes three chapters (7-9) dedicated to
SSU rRNA alignment; chapter 9 is included in this guide as
section~\ref{section:chap9}, but the other chapters may include some
relevant background information as well. 
Other useful references on CMs include
\cite{Eddy94,Durbin98,Eddy02b,NawrockiEddy07,Nawrocki09,KolbeEddy09}. In
addition, the \database{rfam} database 
(\htmladdnormallink{http://rfam.sanger.ac.uk/}{http://rfam.sanger.ac.uk/})
uses \software{infernal} to search for and align
structural RNAs using more than 1000 different CMs, so its
publications may be of interest as well
\cite{Griffiths-Jones03,Griffiths-Jones05,Gardner09}.
\end{comment}

\subsection{How to cite SSU-ALIGN}

Please cite the \sft{infernal} software publication:
\cite{Nawrocki09}, if you find \sft{ssu-align} useful for work that
you publish. Additionally, because \sft{ssu-align}'s seed alignments were
derived from the \db{comparative rna website} we ask that you cite
that database as well: \cite{CannoneGutell02}. 
%Finally, if you'd like to, you
%can also cite my Ph.D. thesis \cite{Nawrocki09b} because it is
%currently the most relevant and comprehensive reference for the
%program.

\subsection{Future development}

This initial 0.1 release of \textsc{SSU-ALIGN} is a prototype. 
Compute time for alignment is roughly one second per full length SSU
sequence. We hope to speed up the program and add additional SSU
profiles in future releases. Bug reports and feature requests are
appreciated.
