\section{Introduction}

\software{ssu-align} identifies and aligns small subunit ribosomal RNA
(SSU rRNA) genes in sequence datasets. It uses models of the primary
sequence and secondary structure conservation of SSU rRNA as inferred
by comparative sequence analysis and confirmed by crystal structure
determination from the Comparative RNA Website (CRW) \cite{Cannone02}.

\subsection{Overview}
\software{ssu-align} uses profile probabilistic models called
covariance models (CMs) for SSU rRNA detection and alignment. It
includes three SSU rRNA models, an archaeal model, a bacterial model,
and a eukaryotic (nuclear) model.  These are the default models used
by the program, but \textsc{ssu-align} can be used with other,
user-generated models as well.

Given an input dataset of unaligned SSU sequences,
\software{ssu-align} goes through two stages. In stage 1, each of the
sequences is locally aligned to each of the five models based on
primary sequence conservation to obtain a score for each sequence to
each model.  The model that gives the highest scoring alignment to
each sequence is the \emph{best-matching model} for that sequence.  In
stage 2, each sequence is aligned to it's best-matching model using
both primary sequence and secondary structure conservation. The end
result is a separate multiple alignment for each model that was the
best-matching model for at least one sequence in the input dataset.

The stage 1 primary sequence alignment is local with respect to both
the model and the sequence. Locality with respect to the model allows
the program to handle sequences that are subsequences of SSU
rRNA that correspond to a specific region of the molecule. Locality
with respect to the sequence allows the program to handle input
sequences that contain extra residues that are nonhomologous to SSU at
the beginning and/or end of the input sequence. In this case, the
local alignment will only correspond to a subsequence of the input
sequence, it is only this subsequence that is aligned to it's
best-matching model in stage 2.

The stage 2 primary sequence and structural based alignment is local
\emph{only} with respect to the model.  This means the subsequence
that survives stage 1 is always included in it's entirety in the final
alignment.

This package can also be used to remove, or prune, unreliable
alignment columns from the final alignments.  This post-processing
step can be useful for removing regions of the alignment that are
are ambiguous and likely include a significant number of errors prior
to using phylogenetic inference tools.

Columns are selected for pruning based on ``confidence estimates'' 
for the aligned residues, which are derived from the probability that
each residue aligns to each column of the alignment. This is discussed
in more detail in the basic tutorial in section~\ref{section:basic}.

The basic tutorial in section~\ref{section:basic} of this guide walks
through an example of using \textsc{ssu-align} to build multiple
alignments and prune those alignments using an input dataset.

This package can also be used to create new SSU models, either for
different phylogenetic clades or for specific regions of the SSU rRNA
molecule, as well as for drawing secondary structure diagrams that
display several different statistics on the alignment (such as levels
of sequence conservation, or frequency of gaps). The advanced
tutorials in section~\ref{section:advanced} of this guide provide
examples of these applications.

\subsection{What is included in this package}

This distribution includes the PERL script \software{ssu-align} and
all the files necessary to compile and run the \software{Infernal}
package (version 1.01) \cite{Nawrocki09}. \software{Infernal} must be
installed in order to run \software{ssu-align}.  It also includes the
three default (archaea, bacteria, and eukarya) SSU models and the
\emph{seed} alignments they were built from. The alignments were based
on SSU structures and alignments from the Comparative RNA Website
(CRW) \cite{Cannone02}. A broader discussion of these models and
alignments and the specific procedure used to create the seed
alignments from the CRW data is explained in \cite{Nawrocki09b}.

Installation of the PERL (Practical Extraction and Report Language,
Larry Wall) interpreter package version 5.0 or later is required to
run the \software{ssu-align} PERL script.

\subsection{What \emph{ssu-align} does not do}

\textsc{ssu-align} only creates alignments; it does not infer trees
from those alignments, nor does it classify sequences beyond reporting
which model in the input CM file they score highest to.

\subsection{Other useful references}

The \software{infernal} user's guide \cite{infernalguide} included in
this package supplements this user's guide. My Ph.D. thesis introduces
and describes CM methods and includes three chapter (7-9) dedicated to
SSU rRNA alignment. Other useful references on CMs include
\cite{Eddy94,Eddy02b,NawrockiEddy07,Nawrocki09,KolbeEddy09}.  The
\database{rfam} database (http://rfam.sanger.ac.uk/) \cite{Gardner09}
uses \software{infernal} to search for and align structural RNAs using
more than 1000 different CMs.

\subsection{Future development}

This release (version 0.1) is a prototype of the \textsc{ssu-align}
software package that only has basic functionality for aligning SSU
sequences. I hope to continue to develop and improve it in the future
by making it faster and generally more useful to the SSU rRNA sequence
analysis community. Bug reports and feature requests are appreciated.
