\section{Description of output files}

This section describes the content of the different output file
formats created by \prog{ssu-align}. Many of the descriptions below
refer to files in the \prog{tutorial/output/} directory. These files
were created by an \prog{ssu-align} run with the command:

\user{ssu-align ../seeds/ssu5-0p1.cm seed-10.fa seed-10
  ../sa-0p1.params}

\subsection{.tab suffixed files}

The \prog{.tab} files are created by the \textsc{infernal} program
\prog{cmsearch} which is called internally by
\prog{ssu-align}. \prog{cmsearch} builds profile HMMs for each model
in the input CM file and scans each input sequence for high scoring
alignments to the HMM\. The \prog{.tab} file lists the locations and
scores of each of these alignments. An example \prog{.tab} file is in 
\prog{tutorial/output/output.tab}. The first part of the file is the
header section:

\begin{sreoutput}
# command:    /Users/eric/infernal-1.0/src/cmsearch --hmm-cW 1.5 -T -1 --no-null3 \\
--toponly --noalign --tab output/output.tab --viterbi output/ssu5-0p1.cm seed-10.fa
# date:       Fri Jun 19 08:23:32 2009
# num seqs:   10
# dbsize(Mb): 0.014568
#
# Pre-search info for CM 1: archaea
#
# rnd  mod  alg  cfg   beta  bit sc cut
# ---  ---  ---  ---  -----  ----------
#   1  hmm  vit  loc      -       -1.00
#
\end{sreoutput}

The first four lines list the \prog{cmsearch} command executed from
within \prog{ssu-align}, the date it was executed, the number of
sequences in the sequence file being searched, and the number of
millions of residues (Mb) in that sequence file. (I manually
wrapped line 1 to make it readable). NOTE ON REVERSE COMPLEMENT!

Next, the pre-search information describing how the search will be
conducted is printed for the first model in the CM file: archaea. 
The \prog{rnd}, \prog{mod}, \prog{alg} and \prog{cfg} columns report
that the first and only round of searching will be conducted with a
profile \prog{hmm} using the \prog{vit}erbi algorithm for scoring
\prog{loc}al alignments to the model. (The viterbi algorithm is
briefly explained in the ``Background'' section). The \prog{beta}
column is irrelevant when searching with an HMM. The \prog{bit sc cut}
column shows the minimum score threshold, alignments scoring above
\prog{-1.00} bits will be reported. MORE HERE EXPLAINING WHY -1?

After the pre-search information comes four lines of column headings
followed by data lines reporting high scoring local alignments. One
line is printed per alignment, or \emph{hit} (it is possible to see
more than one hit per sequence). Remember these are scores against the
archaea model.

\begin{sreoutputtiny}
#                                                       target coord   query coord                         
#                                             ----------------------  ------------                         
# target name                                      start        stop  start   stop    bit sc   E-value  GC%
# ------------------------------------------  ----------  ----------  -----  -----  --------  --------  ---

  00841::Escherichia_coli::J01695                     54        1540      1   1508    622.99         -   55
  00116::Pyrococcus_horikoshii::AP000001               1        1497      1   1508   2215.49         -   66
  00628::Saccharomyces_cerevisiae.::U53879          1109        1797      1   1508    196.51         -   49
  00230::Oxytricha_granulifera::X53486              1096        1775      1   1508    191.89         -   50
  00366::Mitochondrion_Mus_musculus::J01420          831         927      1   1508     15.94         -   38
  00173::chloroplast_Zea_mays_maize_::Z00028          54        1488      1   1508    703.54         -   56
  00004::Nanoarchaeum_equitans::AJ318041               1        1500      1   1508   1916.64         -   68
  01058::Mycobacterium_tuberculosis.::Z83862          57        1533      1   1508    702.11         -   58
  00569::Mitochondrion_Gallus_gallus::X52392         856         948      1   1508      3.96         -   47
  00127::Nostoc_muscorum::X59559                      10        1487      1   1508    697.37         -   54
\end{sreoutputtiny}

Here's a brief description of each column:

\begin{wideitem}
\item[\emprog{target name}] the name of the target sequence.

\item[\emprog{start (target coord)}] first sequence residue in the alignment

\item[\emprog{end (target coord)}] final sequence residue in the alignment
\end{wideitem}

The \prog{begin} and \prog{end} columns under \prog{query coord} are
uninformative for HMM searches like those performed by
\prog{ssu-align}. (\prog{cmsearch} will print informative numbers in
these columns when the CM is used for a search.) With HMM searches the
\prog{begin} column will always be \prog{1} and the \prog{end} column
will always be the final position of the model.

\begin{wideitem}

\item[\emprog{bit sc}] the bit score of the HMM alignment of
  target sequence residues \prog{begin} to \prog{end}.

\item[\emprog{E-value}] this will always read \prog{-}. It would
  include an E-value of the bit score if the model had been calibrated
  with \prog{cmcalibrate}. Currently, \prog{ssu-align} does not use
  calibrated models, mainly because they are most useful for
  identifying remotely homologous structural RNAs using much models of
  much smaller RNAs, such as transfer RNA. See the \textsc{infernal}
  user's guide for more information.

\item[\emprog{GC\%}] the percent of the residues in the aligned
  sequence that are either \prog{G} or \prog{C}. This is largely
  irrelevant for SSU rRNA, but is sometimes useful when using
  \prog{cmsearch} with smaller models.
\end{wideitem}

\subsection{.hits.list suffixed files}

The \prog{.hits.list} files are simple files created for each model
that list the sequences that were best-matches to that model, and as a
result were aligned to that model in stage 2. If a model has zero
sequences for which it is the best-matching model, this file will not
be created. The file contains no new information absent from the
\prog{.tab} file and is only created as a convenience, because it is
cumbersome to determine which model gave the highest score to a
particular sequence in the \prog{tab} file.  For model \emph{x}, the
\prog{hits.list} file contains four columns providing four pieces of
data for each sequence whose best-matching model was \emph{x}. For
example, take a look at the \prog{output/output.archaea.hits.list}
file:

\begin{sreoutput}
# List of 2 subsequences to align to CM: archaea
# Created by ssu-align.pl.
#
# target name                              start    stop     score
# --------------------------------------  ------  ------  --------
  00116::Pyrococcus_horikoshii::AP000001       1    1497   2215.49
  00004::Nanoarchaeum_equitans::AJ318041       1    1500   1916.64
\end{sreoutput}

The four columns correspond to those of the same name in the {.tab}
suffixed files as explained above: the \prog{target name},
\prog{start} and \prog{stop} (which correspond to the target sequence
coordinates), and \prog{score}, which is the primary sequence HMM
score assigned to the sequence by model \emph{x}: the HMM score of all
other models to this sequence is less than this score.

\subsection{.hits.fa suffixed files}
This is a fasta formatted sequence file containing the sequences
listed in the corresponding \prog{hits.list} file. For example, the
file \prog{output/output.archaea.hits.fa} contains the two sequences
listed in \prog{output/output.archaea.hits.list}. These sequences were
copied from the original fasta sequence file \prog{seed-10.fa} that
was used as input to \prog{ssu-align}. Only the residues from
positions \prog{start} to \prog{stop} (as listed in the
\prog{hits.list} file) were copied, so sometimes the sequences in the
\prog{hits.fa} file will be subsequences of those from the original
file.  The \prog{ssu-align} script uses the \prog{hits.fa} files it
creates as input to the \prog{cmalign} program, which it calls
internally to generate the alignments.

\subsection{.cmalign suffixed files}

The \prog{.cmalign} files are the standard output created by the
\textsc{infernal} program \prog{cmalign} which is called internally
during stage 2 of the \prog{ssu-align} script. There is one such file
created for each model that was the best-matching model to at least
one sequence in \prog{ssu-align}'s search stage. Take a look at the
file \prog{output/output.archaea.cmalign}. The first part of the file
is the header section:

\begin{sreoutput}
# cmalign :: align sequences to an RNA CM
# INFERNAL 1.0 (January 2009)
# Copyright (C) 2009 HHMI Janelia Farm Research Campus
# Freely distributed under the GNU General Public License (GPLv3)
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# command: /Users/eric/infernal-1.0/src/cmalign -p --mxsize 4096 --sub \\
-o output/output.archaea.stk output/ssu5-0p1.archaea.m1.cm output/output.archaea.hits.fa
# date:    Fri Jun 19 08:24:02 2009
\end{sreoutput}

This section includes the program version used, the copyright
information, the command used to execute \prog{cmalign}, and the date
of execution.

Next comes information on alignment parameters used by the program:

\begin{sreoutput}
#
# cm name                    algorithm  config  sub  bands     tau
# -------------------------  ---------  ------  ---  -----  ------
# archaea                      opt acc  global  yes    hmm   1e-07
\end{sreoutput}

The \prog{cm name} column reports the name of the model used for
alignment. \prog{algorithm} gives the name of the algorithm, in this
case \prog{opt acc} stands for \emph{optimal accuracy}. This is a
dynamic programming algorithm that returns the maximal posterior
labelling of all emitted residues in the alignment CITE(HolmesXX) (for
more information see the ``Background'' section). The next two
columns, \prog{config} and \prog{sub}, read \prog{global} and
\prog{yes} respectively, which tells us the program will first predict
the start and end points of the alignment to the model using an HMM
(the \prog{sub yes} part) and then align the region of the model that
spans from start to end \emph{globally} to the sequence (the
\prog{global} part). In this case, \emph{global} alignment means that
the program is forced to align the full model region from start to end
to the sequence, e.g. it is \emph{not} allowed to skip large parts of
the model without large score penalties as it would if \emph{local}
alignment was being performed. The \prog{bands} column tells us that
bands (constraints) from an HMM alignment will be used to accelerate
alignment to the CM. This is explained more in the ``Background''
section. The \prog{tau} column reports the probability loss allowed
when computing the HMM bands. In this case, \prog{1e-07} probability
mass is allowed outside each band (again, see ``Background'' for more
information).

The next section includes per-sequence information on the alignment
that was created:

\begin{sreoutputtiny}
# seq idx  seq name                                  len     total    struct  avg prob      elapsed
# -------  --------------------------------------  -----  --------  --------  --------  -----------
        1  00116::Pyrococcus_horikoshii::AP000001   1497   2316.02    176.14     1.000  00:00:03.00
        2  00004::Nanoarchaeum_equitans::AJ318041   1500   2094.59    265.20     0.996  00:00:02.90
\end{sreoutputtiny}

We'll go through each of these columns:


\begin{wideitem}
\item[\emprog{seq idx}] the index of the sequence in the file.

\item[\emprog{seq name}] the name of the sequence.

\item[\emprog{len}] length of the sequence; the full sequence is
  aligned, no trimming of ends is permitted, as it was in the search
  stage with \prog{cmsearch}.

\item[\emprog{total}] the bit score of the CM alignment. For more
  information, see the ``Background'' section.

\item[\emprog{struct}] the secondary structure score component of the
  \prog{total} bit score. These are the added bits that are due solely
  to the modelling of the consensus secondary structure of the
  molecule by the CM\. For more information, see the ``Background'' section.

\item[\emprog{avg prob}] the average posterior labeling, or confidence
  estimate, of the aligned residues. The higher this value is the less
  ambiguous and more well-defined the alignment is. The highest this
  can possibly be is \prog{1.000}, which means very nearly 100\% of
  the probability mass of the alignment to the model is contained in
  the single, optimally accurate alignment that was reported by the
  program. In other words, the reported alignment receives a
  significantly higher score than any other alternative alignment. The
  program derives this value by evaluating the score of every possible
  alignment (consistent with the HMM bands) of the sequence to the
  model, and comparing the best, optimal score versus all of the
  rest. For more information, see the ``Background'' section.

\item[\emprog{elapsed}] the amount of actual time (wall time) it took
  the program to align this sequence. In general, less well defined
  alignments with lower \prog{avg prob} will take longer than more
  well-defined ones. This is because the HMM bands are usually tighter
  and act as stricter constraints to the CM alignment when the
  alignment is well defined. Tighter bands lead to quicker alignments
  because the number of possible alignments to the CM that must be
  considered is less.
\end{wideitem}

\subsection{.stk suffixed files}
The \prog{.stk} suffixed files are Stockholm-formatted alignment
files. These are the alignments generated by \prog{cmalign}. The
statistics in the \prog{.cmalign} suffixed files correspond to these
alignments. One alignment is created for each model that was the
best-matching model to at least one sequence in \prog{ssu-align}'s
search stage. An explanation of Stockholm alignments can be found in
the ``Basic tutorial'' section.

\subsection{.scores suffixed files}

The \prog{.scores} file are meant to be a useful summary file for each
run of \prog{ssu-align}. They contain various statistics from each of
the other output files for every sequence in the original input
sequence file. 

Take a look at the file \prog{tutorial/output/output.scores}

\begin{sreoutputtinywide}
#                                                                     best-matching model                   second-best model  
#                                                      --------------------------------------------------  --------------------
#     idx  sequence name                               model name    beg   end    CM sc   struct   HMM sc  model name    HMM sc  HMMdiff
# -------  ------------------------------------------  -----------  ----  ----  -------  -------  -------  -----------  -------  -------
        1  00841::Escherichia_coli::J01695             bacteria        1  1542  2398.43   303.43  2125.65  chloroplast  1427.15   698.50
        2  00116::Pyrococcus_horikoshii::AP000001      archaea         1  1497  2316.02   176.14  2215.49  bacteria      601.72  1613.77
        3  00628::Saccharomyces_cerevisiae.::U53879    eukarya         1  1800  2753.53   162.93  2627.74  archaea       196.51  2431.23
        4  00230::Oxytricha_granulifera::X53486        eukarya         1  1778  2636.03   143.04  2515.92  archaea       191.89  2324.03
        5  00366::Mitochondrion_Mus_musculus::J01420   metamito        1   956  1259.62    87.37  1219.28  -                  -        -
        6  00173::chloroplast_Zea_mays_maize_::Z00028  chloroplast     2  1490  2381.70   118.29  2305.29  bacteria     1781.63   523.66
        7  00004::Nanoarchaeum_equitans::AJ318041      archaea         1  1500  2094.59   265.20  1916.64  bacteria      480.94  1435.70
        8  01058::Mycobacterium_tuberculosis.::Z83862  bacteria        2  1537  2324.10   238.30  2140.25  chloroplast  1410.10   730.15
        9  00569::Mitochondrion_Gallus_gallus::X52392  metamito        1   976  1129.62   123.26  1061.76  -                  -        -
       10  00127::Nostoc_muscorum::X59559              chloroplast     2  1489  2263.17   211.15  2101.91  bacteria     2017.96    83.95
\end{sreoutputtinywide}

There are four rows containing column headings prefixed with
\prog{\#}. Then there are 10 data rows, one for each sequence in the
input sequence file \prog{tutorial/seed-10.fa}. Data rows are
separated into 12 columns:

\begin{wideitem}
\item[\emprog{idx}] the index of the sequence in the file.

\item[\emprog{sequence name}] the name of the sequence.
\end{wideitem}

The next 6 columns all describe the \emph{best-matching} model for the
sequence. This is the model that assigned the highest primary
sequence-based local profile HMM alignment score to the sequence. 
If no model aligned the sequence with a score higher than the minimum
threshold of $100$ bits then the sequence was skipped and not
aligned, and all these columns will read \prog{-}. (Note: the minimum
bit score threshold value can be changed to \prog{<x>} using the
\prog{ssu-align} command-line option \prog{-b <x>}). 

\begin{wideitem}
\item[\emprog{model name}] name of best-matching model.

\item[\emprog{beg}] first sequence residue index in the maximal
  scoring local HMM alignment to the best-matching model.

\item[\emprog{end}] final sequence residue index in the maximal
  scoring local HMM alignment to the best-matching model.

\item[\emprog{CM sc}] the CM bit score for the best-matching model
  assigned to the sequence from positions \prog{beg} to \prog{end}.

\item[\emprog{struct}] the number of extra bits included in the
  CM bit score that are dervied from the secondary structure component
  of the model.

\item[\emprog{HMM sc}] the HMM bit score for the local alignment of
  the best-matching model to the sequence from positions \prog{beg} to
  \prog{end}.
\end{wideitem}

The ``Background'' section contains a more detailed discussion of bit
scores.

The next 2 columns describe the \prog{second-best model}. This is the
model that assigned the second-highest primary sequence-based local 
profile HMM alignment score to the sequence. If only one models score
exceeded the minimum of $100$ bits then these columns will each read
``\prog{-}''.

\begin{wideitem}
\item[\emprog{model name}] name of second-best-matching model.

\item[\emprog{HMM sc}] the HMM bit score for the local alignment of
  the second-best-matching model to the sequence. This alignment was
  not necessarily from \prog{beg} to \prog{end} (those were the
  coordinates of the alignment to the best-matching model). The
  sequence coordinates of the second-best model's alignment can be
  found in the file \prog{output.tab}.
\end{wideitem}

The final column, \prog{HMMdiff}, reports the score difference between
the best-matching model HMM alignment and the second-best matching
model HMM alignment. This is included because it is an indication of
how clearly homologous the sequence is to the best-matching model
instead of the second-best-matching model. The higher this score
difference is the more obvious it is that the sequence falls within
the sequence diversity represented by the best-matching model.
Sequences that are phylogenetically novel and do not obviously match
any single model much better than any other one will have score
differences in this column.

\subsection{.log suffixed files}

The \prog{.log} files include the text reported to the screen
(standard output) by \prog{ssu-align}. These files serve as a
reference to remind the user how the \prog{ssu-align} script was run
(parameters, input file names, etc.). 



