\section{Tutorial}
\label{section:tutorial}

Here is a tutorial walk-through of some small projects with
\sft{ssu-align}. To follow along with this tutorial, move into the
\prog{tutorial/} subdirectory of the {ssu-align-0.1/} directory where
you unpacked and built the package.
The instructions in this tutorial assume that you have already
installed the package (see the Installation section)
and all of the \sft{ssu-align} executable programs
in your \$PATH. For example, you should be able to run
\prog{ssu-align} by simply typing \prog{ssu-align}.

\begin{comment}
\subsection{Files used in this tutorial}

In the first section of this tutorial we'll use the following files in
the \prog{tutorial} directory:

  \begin{sreitems}{}
  \item[\prog{seed-15.fa}] a sequence file containing
    fifteen SSU rRNA sequences, created specifically for use in this
    tutorial. These are full or partial sequences from the archaeal,
    bacterial and eukaryotic default \sft{crw} seed alignments used to
    build the default \software{ssu-align} models.
  \end{sreitems}
\end{comment}

\subsection{Creating, masking and visualizing SSU alignments}

We'll use a small dataset to demonstrate how the package works.
The file \prog{seed-15.fa} contains five archaeal sequences, five
bacterial sequences and five eukaryotic sequences from the
\sft{ssu-align} v0.1 seed alignments. These seed alignments
were derived from alignments from the \db{CRW} database
\cite{CannoneGutell02} as described in section~\ref{section:chap9}.

Pretend that \prog{seed-15.fa} is a set of SSU sequences obtained from
an environmental sampling survey and that we want to analyze. First,
we can run the \prog{ssu-align} program to classify each sequence to
its domain of life, and create an alignment for each domain:

\user{ssu-align seed-15.fa myseqs}\\

As you can see, the \sft{ssu-align} script takes two command line
arguments. The first is the target sequence file. The second is the
name of a directory that \sft{ssu-align} will create and place its
output files into. This directory should not yet exist. 

The program will first print a header describing the program version
used, command used, current date, and some other information. 
The following information printed to the screen:

\begin{sreoutput}
# ssu-align :: align SSU rRNA sequences
# SSU-ALIGN 0.1 (May 2010)
# Copyright (C) 2010 HHMI Janelia Farm Research Campus
# Freely distributed under the GNU General Public License (GPLv3)
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# command: ssu-align seed-15.fa myseqs
# date:    Wed May  5 13:17:20 2010
#
# Validating input sequence file ... done.
#
# Stage 1: Determining SSU start/end positions and best-matching models...
\end{sreoutput}

In stage 1, the program scans the input sequences with each of the
three default SSU models. This has two purposes.  First, it classifies
each sequence by determining which model in the input CM file is its
``best-matching'' model, defined as
the model that gives the sequence the highest primary sequence-based
alignment score using a profile HMM. Secondly, it
defines the start and end points of the SSU sequences based on the
best-matching model's alignment.

Stage 1 takes about 20 seconds on this dataset (on an Intel Xeon 3.0
GHz processor, which I'll use for all example runs in this
guide). When it finishes you'll see: 

\begin{sreoutput}
# Stage 1: Determining SSU start/end positions and best-matching models...
#
# output file name         description                                
# -----------------------  -------------------------------------------
  myseqs.tab               locations/scores of hits defined by HMM(s)
  myseqs.archaea.hitlist   list of sequences to align with archaea CM
  myseqs.archaea.fa              5 sequences to align with archaea CM
  myseqs.bacteria.hitlist  list of sequences to align with bacteria CM
  myseqs.bacteria.fa             5 sequences to align with bacteria CM
  myseqs.eukarya.hitlist   list of sequences to align with eukarya CM
  myseqs.eukarya.fa              5 sequences to align with eukarya CM
\end{sreoutput}

This lists and briefly describes the seven new files the script created
in the newly created \prog{myseqs/} subdirectory of the current directory.
The content and format of these files are described
in detail in section~\ref{section:output}. For now a brief explanation
should be sufficient. The first file \prog{myseqs.tab} is output from
\sft{infernal}'s \prog{cmsearch} program. The other six files are
model-specific: two files for each model that was the best-matching
model for at least one sequence in the input target sequence file
\prog{seed-15.fa}. The \prog{.hitlist} suffixed files contain a list
of the sequences that match best to the model, and the \prog{.fa}
suffixed files are those actual sequences. If any of the models had
not been the best-matching model to at least one target sequence,
there would be no \prog{.hitlist} or \prog{.fa} files for that
model.

The program will now proceed to stage 2, the alignment stage. This
stage serially progresses through each model that was the
best-matching model for at least one sequence and uses the model to
align the best-matching sequences. The alignments are computed by scoring
a combination of both sequence and secondary structure conservation,
as opposed to the scoring in stage one which only used sequence
conservation. As the alignment to each model finishes, two new lines
of text, one for each of two newly created files, will appear on the
screen. For this example, alignment to all three models takes about 20
seconds. When it finishes you'll see:

\begin{sreoutput}
# Stage 2: Aligning each sequence to its best-matching model...
#
# output file name         description
# -----------------------  ---------------------------------------
  myseqs.archaea.stk       archaea alignment
  myseqs.archaea.cmalign   archaea cmalign output
  myseqs.archaea.ifile     archaea insert info
  myseqs.bacteria.stk      bacteria alignment
  myseqs.bacteria.cmalign  bacteria cmalign output
  myseqs.bacteria.ifile    bacteria insert info
  myseqs.eukarya.stk       eukarya alignment
  myseqs.eukarya.cmalign   eukarya cmalign output
  myseqs.eukarya.ifile     eukarya insert info
  myseqs.scores            list of CM/HMM scores for each sequence
\end{sreoutput}

The newly created alignments are the \prog{.stk} suffixed files. These
were created by \sft{infernal}'s \prog{cmalign} program. The
\prog{.cmalign} and \prog{.ifile} suffixed files were also output by
\prog{cmalign}. As in stage 1, these files were created in the
\prog{./myseqs/} subdirectory. 

\subsubsection{Description of alignments}

Section~\ref{section:output} contains more information
on \sft{ssu-align} output files, but for now we'll focus only on the new
alignments.  Take a look at the archaeal alignment we just created in
\prog{myseqs/myseqs.archaea.stk}.

This alignment includes consensus secondary structure annotation and
is in \emph{Stockholm format}. 
Stockholm format, the native alignment format used by \sft{hmmer} and
\sft{infernal} and the \db{pfam} and \db{rfam}
databases, is documented in detail in the \sft{infernal} user's
guide which is included as a PDF in \prog{infernal-1.01/}

For now, what you need to know about the key features of the alignment file is:
\begin{itemize}

\item The alignment is in an interleaved format, like other
  common alignment file formats such as \sft{clustalw}.
  Lines consist of a name, followed by an aligned sequence;
  the alignment is split into blocks separated by blank lines.

\item Gaps are indicated by the characters ., \_, -, or \verb+~+.
  Many SSU alignments will have large regions of 100\% gaps at the
  beginning and ends of the alignment. 
  This will happen if the sequences are 
  partial SSU sequences, such as those  obtained with PCR
  primers that target well conserved regions within the SSU
  molecule.

\item Special lines starting with {\small\verb+#=GR+} followed by a
  sequence name and then {\small\verb+PP+} contain posterior
  probabilities for each aligned nucleotide for the sequence they
  correspond to.  These are confidence estimates in the correctness of
  the alignment.  Figure~\ref{fig:ambiguity} in
  section~\ref{section:chap9} provides an example of confidence
  estimates and posterior probabilities.  Characters in PP rows have
  12 possible values: "0-9", "*", or ".". If ".", the position
  corresponds to a gap in the sequence. A value of "0" indicates a
  posterior probability of between 0.0 and 0.05, "1" indicates between
  0.05 and 0.15, "2" indicates between 0.15 and 0.25 and so on up to
  "9" which indicates between 0.85 and 0.95. A value of "*" indicates
  a posterior probability of between 0.95 and 1.0. Higher posterior
  probabilities correspond to greater confidence that the aligned
  nucleotide belongs where it appears in the alignment.  These
  confidence estimates can be used to mask the alignment to remove
  columns with significant fractions of ambiguously aligned nucleotides
  as demonstrated below.

\item A special line starting with {\small\verb+#=GC SS_cons+}
  indicates the secondary structure consensus. Gap characters annotate
  unpaired (single-stranded) columns. Base pairs are indicated by any
  of the following pairs: \verb+<>+, \verb+()+, \verb+[]+, or
  \verb+[]+.

\item A special ``RF'' line starting with {\small\verb+#=GC RF+}
  indicates the consensus, or ReFerence, model. Gaps in the RF line
  are \emph{insert} columns, where at least 1 sequence has at least 1
  inserted nucleotide between two consensus positions. Uppercase nucleotides
  in the RF line are well conserved positions in the model; lowercase
  nucleotides are less well conserved.
\end{itemize}

\subsection{Masking (removing columns from) alignments}

If your goal is to use a phylogenetic inference program to build trees
from alignments created by \sft{ssu-align}, you may want to mask
out columns of the alignment that may include misaligned nucleotides
first, and then only run the inference on the remaining columns where
you're confident the alignment is correct. 

\begin{comment}
SSU alignments are commonly used for phylogenetic inference to
characterize the diversity of microorganisms in an environmental
sample. Because alignment errors confound phylogenetic inference, 
it is first recommended to remove columns from, or mask, SSU
alignments to remove regions that are likely to contain at least some
errors. 
\end{comment}

The \prog{ssu-mask} program uses the posterior probabilities
(PP values) in the alignments to determine which alignment columns
contain a significant fraction of nucleotides that are aligned with
low confidence. It takes a single command-line argument, the name of
the directory created by \prog{ssu-align}. The directory must exist
within the current working directory. To run it for our example, do: 

\user{ssu-mask myseqs} 

As with \prog{ssu-align}, the program will print information to the
screen about what it is doing and the files it is creating:

\newpage 

\begin{sreoutput}
# Masking alignments based on posterior probabilities...
#
#                                                    mask    
#                                                ------------
# file name                 in/out  type  #cols  incl.  excl.
# ------------------------  ------  ----  -----  -----  -----
  myseqs.archaea.stk         input   aln   1511      -      -
  myseqs.archaea.mask       output  mask   1508   1449     59
  myseqs.archaea.mask.pdf   output   pdf   1508   1449     59
  myseqs.archaea.mask.stk   output   aln   1449      -      -
#
  myseqs.bacteria.stk        input   aln   1597      -      -
  myseqs.bacteria.mask      output  mask   1582   1499     83
  myseqs.bacteria.mask.pdf  output   pdf   1582   1499     83
  myseqs.bacteria.mask.stk  output   aln   1499      -      -
#
  myseqs.eukarya.stk         input   aln   2009      -      -
  myseqs.eukarya.mask       output  mask   1881   1634    247
  myseqs.eukarya.mask.pdf   output   pdf   1881   1634    247
  myseqs.eukarya.mask.stk   output   aln   1634      -      -
\end{sreoutput}

The 'file name' column includes file names, either input or output
files, as listed in the 'in/out' field, with type specified by the
'type' column: 'aln' for alignment, 'mask' for mask file, and 'pdf' or
'ps' for structure diagram file. The 'cols' column gives the number of
columns for the file. The 'incl.' and 'excl.' columns list the number
of consensus columns of the alignment that are included (not removed)
and excluded (removed), respectively, by the mask. For example, the
\prog{myseqs.archaea.stk} input alignment was initially 1511 total columns,
1508 of which were consensus columns; 59 of these 1508 were deemed
unreliable and removed by the mask based on the PP values in those
columns, and the remaining 1449 were not removed. The alignment file
\prog{myseqs.archaea.mask.stk} was created that includes only these 1449
columns, and a structure diagram showing the positions of the 1449
columns in the context of the full archaeal consensus structure was
written to \prog{myseqs.archaea.mask.pdf} \footnote{In your hands, a
postscript file \prog{myseqs.archaea.mask.ps} may be created instead
of this pdf file. This will happen if you do not have the
\prog{ps2pdf} program installed and in your PATH. See the manual
page for \prog{ssu-mask} for more information.}

After running this command, the three files with \prog{.mask.stk}
suffixes in the \prog{myseqs} directory are your masked
alignments. \textsc{ssu-align} has high confidence that the columns in
these alignments contain very few errors. 

\subsubsection{using precalculated masks for consistent masking of
  multiple datasets}

In the \prog{ssu-mask} example above, we calculated a mask
specifically based on our \prog{myseqs} alignments. Given a different
dataset of SSU sequences, the generated masks would very likely be
different (i.e. exclude different columns). This
per-alignment-specificity can be undesirable. For example, if you'd
like to directly compare alignments from multiple runs of
\textsc{ssu-align} you probably want to make sure all of the
alignments being compared contain an identical set of consensus
positions \footnote{This would also allow the alignments to be combined
easily; see the \prog{ssu-merge} manual page.}. \textsc{ssu-align}
contains a default, precalculated mask for each of the three
models. These masks were derived from very large alignments as
described in section X (TO DO). As we'll see in the next example, you
can use these masks by supplying the \prog{-d} option to
\prog{ssu-mask}. We'll also use the \prog{--key-out} option to add the
string 'default' to the names of the output files so we do not
overwrite the files we created in the previous example. This option
enables you to save multiple sets of alignments and masks in a single
directory:

\user{ssu-mask -d --key-out default myseqs}

\begin{sreoutput}
# Masking alignments using pre-existing masks...
#
#                                                    mask    
#                                                ------------
# file name                 in/out  type  #cols  incl.  excl.
# ------------------------  ------  ----  -----  -----  -----
  myseqs.archaea.stk         input   aln   1511      -      -
  archaea-0p1.mask           input  mask   1508   1376    132
  myseqs.archaea.mask.pdf   output   pdf   1508   1376    132
  myseqs.archaea.mask.stk   output   aln   1376      -      -
#
  myseqs.bacteria.stk        input   aln   1597      -      -
  bacteria-0p1.mask          input  mask   1582   1376    206
  myseqs.bacteria.mask.pdf  output   pdf   1582   1376    206
  myseqs.bacteria.mask.stk  output   aln   1376      -      -
#
  myseqs.eukarya.stk         input   aln   2009      -      -
  eukarya-0p1.mask           input  mask   1881   1343    538
  myseqs.eukarya.mask.pdf   output   pdf   1881   1343    538
  myseqs.eukarya.mask.stk   output   aln   1343      -      -
\end{sreoutput}

The output is similar to the previous example, but note that the
\prog{.mask} suffixed files were input rather than output (these files
are placed in your \$SSUALIGNDIR during installation), and the number
of columns included and excluded has changed.

There are several other options that can be supplied to
\prog{ssu-mask}, including \prog{-f} and \prog{-k} which gives you the
ability to use your own precalculated masks. See the \prog{ssu-mask}
manual page for more information. 

\subsubsection{converting Stockholm alignments to FASTA format}
With the ambiguously aligned regions of the alignment removed, you may
want to use the alignments as input to a phylogenetic inference
program, but not many of those programs can accept Stockholm formatted
alignments. You can convert the Stockholm alignments to aligned FASTA
using the \prog{ssu-mask} program by specifying the \prog{--stk2afa}
option on the command line:

\user{ssu-mask --stk2afa myseqs}

After running, three \prog{.afa} suffixed files will have been created
in the \prog{myseqs} directory.


\input{tutorial-draw}

\input{tutorial-trunc}

\input{tutorial-prep}

\input{tutorial-merge}







