\section{Tutorial}
\label{section:tutorial}

Here is a tutorial walk-through of some small projects with
\sft{ssu-align}. To follow along with this tutorial, move into the
\prog{ssu-align-0.1} subdirectory of \prog{infernal-1.01/} directory
where you unpacked and built the \sft{infernal} package. Once there
create a new directory, something like \prog{my-tutorial} and move
into it. The instructions in this tutorial assume that you have the
\sft{ssu-align} perl script and \sft{easel} programs in your path. For
example, you should be able to run \prog{ssu-align} and
\prog{esl-alimanip} by simply typing \prog{ssu-align} or
\prog{esl-alimanip} at the UNIX prompt.  These instructions also
assume that you have setup your local
\prog{ssu-align-0.1/sa-0p1.params} file to include your local
paths. Instructions on how to do all of this are included in the
Installation section.

\begin{comment}
The instructions in this tutorial assume that you
have the \sft{ssu-align} perl script and \sft{easel} programs in your
path. For example, you should be able to run \emph{version 1.01} of
\prog{ssu-align} and \prog{esl-alimanip} by simply typing
\prog{ssu-align} or \prog{esl-alimanip} at the UNIX prompt.  The
instructions also assume that you have setup your local
\prog{sa-0p1.params} to include your local paths to the executables.
Instructions on how to do all of this are included in the the
Installation section. Frequently, I will refer to files included in
the distrubution. To specify their locations, I will use the strings
\prog{\$INFDIR} and \prog{\$SADIR} to refer to the top-level
\prog{infernal-1.01} directory and its \prog{ssu-align-0.1}
subdirectly respectively.
\end{comment}
%One final note: for many of the small
%tutorials below I ask you to copy files into new directories before
%performing the tutorials, this is not good practice, you could just 

\subsection{Files used in this tutorial}

In the first section of this tutorial we'll use the following files:

  \begin{sreitems}{}
  \item[\prog{ssu-align-0.1/seeds/ssu3-0p1.cm}] A covariance model (CM) file that
    defines three SSU rRNA CMs: an archael model, a bacterial model,
    and a eukaryotic model. These are the three default models used by
    \sft{ssu-align}. More information can be found on these models
    in section~\ref{section:chap9}.
  \item[\prog{ssu-align-0.1/tutorial/seed-15.fa}] a sequence file containing
    fifteen SSU rRNA sequences, created specifically for use in this tutorial.
  \item[\prog{ssu-align-0.1/sa-0p1.params}] A file containing paths to
    \sft{infernal} executable files that \sft{ssu-align} needs
    to run. You will likely need to change
    these paths to point to where you've installed the \prog{cmsearch}
    and \prog{cmalign} programs as noted in the Installation section.
  \end{sreitems}

Other files will be explained as needed in the later sections of the tutorial.
%Create a new directory that you can work in, and copy the three
%files listed above there. I'll assume for the following examples that you've
%installed the \prog{ssu-align} PERL script in your path; if not, you'll
%need to give a complete path name to the script (e.g. something like
%\newline
%\prog{/usr/people/nawrocki/infernal-1.01/ssu-align-0.1/ssu-align} 
%instead of just \prog{ssu-align}).

\subsection{Creating structural SSU alignments with \sft{ssu-align}}

First, we'll go through how to use the program for
it's most basic and fundamental purpose, creating multiple
alignments of SSU rRNA sequences. 

The file \prog{seed-15.fa} contains five archaeal sequences, five
bacterial sequences and five eukaryotic sequences from the
\sft{ssu-align} v0.1 seed alignments. These seed alignments
were derived from alignments from the \db{CRW} database
\cite{CannoneGutell02} as described in section~\ref{section:chap9}.
This sequence file was created to allow a quick demonstration of
\sft{ssu-align}'s ability to create structurally annotated multiple
alignments of SSU sequences. To run this example, 
execute the following command:

\scriptuser{ssu-align ../seeds/ssu3-0p1.cm ../tutorial/seed-15.fa seed-15 ../sa-0p1.params}\\

As you can see, the \sft{ssu-align} script takes four command line
arguments. The first is a CM file. The default CM file
\prog{ssu3-0p1.cm} contains three CMs: an archaeal SSU model, a
bacterial SSU model and a eukaryotic SSU model. The second argument is
the target sequence file to search (this must be in FASTA format). The
third is a string that is used as the basis for naming output
files. In this case, the script will create a new directory called
\prog{seed-15} where all output files will be created; \prog{seed-15}
will also be included as a prefix to the names of all output files.

\newpage

The program will first print a header describing the program version
used, command used, current date, and some other information. 
The following information printed to the screen:

\begin{sreoutput}
# ssu-align :: define and align SSU rRNA sequences
# SSU-ALIGN 0.1 (June 2009)
# Copyright (C) 2009 HHMI Janelia Farm Research Campus
# Freely distributed under the GNU General Public License (GPLv3)
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# command: wd-1p0-branch/ssu-align-0.1/ssu-align -F ssu3-0p1.cm seed-15.fa seed-15 sa-0p1.params
# date:    Thu Aug 20 06:32:04 2009
#
# Stage 1: Determining SSU start/end positions and best matching models.
\end{sreoutput}

In stage 1, the program scans the input sequences with each of the
three models in the CM file \\ \prog{ssu3-0p1.cm}. This has two
purposes.  First, it classifies each sequence by determining which
model in the input CM file is its ``best-matching'' model, defined as
the model that gives the sequence the highest primary sequence-based
alignment score using a profile HMM. Secondly, it
defines the start and end points of the SSU sequences based on the
best-matching model's alignment.

Stage 1 takes about 20 seconds on this dataset (on an Intel Xeon 3.0
GHz processor, which I'll use for all example runs in this
guide). When it finishes you'll see: 

\begin{sreoutput}
# Stage 1: Determining SSU start/end positions and best matching models.
#
# output file name            description                                
# --------------------------  -------------------------------------------
  seed-15.tab                 locations/scores of hits defined by HMM(s)
  seed-15.archaea.hits.list   list of sequences to align with archaea CM
  seed-15.archaea.hits.fa           5 sequences to align with archaea CM
  seed-15.bacteria.hits.list  list of sequences to align with bacteria CM
  seed-15.bacteria.hits.fa          5 sequences to align with bacteria CM
  seed-15.eukarya.hits.list   list of sequences to align with eukarya CM
  seed-15.eukarya.hits.fa           5 sequences to align with eukarya CM
\end{sreoutput}

This lists and briefly describes the seven new files the script created
in a newly created subdirectory of the current working dir called
\prog{seed-15/}. The content and format of these files are described
in detail in section~\ref{section:output}. For now a brief explanation
should be sufficient. The first file \prog{seed-15.tab} is output from
\sft{infernal}'s \prog{cmsearch} program. The other six files are
model-specific: two files for each model that was the best-matching
model for at least one sequence in the input target sequence file
\prog{seed-15.fa}. The \prog{.hits.list} suffixed files contain a list
of the sequences that match best to the model, and the \prog{.hits.fa}
suffixed files are those actual sequences. If any of the models had
not been the best-matching model to at least one target sequence,
there would be no \prog{.hits.list} or \prog{.hits.fa} files for that
model.

The program will now proceed to stage 2, the alignment stage. This
stage serially progresses through each model that was the
best-matching model for at least one sequence and uses the model to
align the best-matching sequences. The alignments are computed by scoring
a combination of both sequence and secondary structure conservation,
as opposed to the scoring in stage one which only used sequence
conservation. As the alignment to each model finishes, two new lines
of text, one for each of two newly created files, will appear on the
screen. For this example, alignment to all three models takes about 20
seconds. When it finishes you'll see:

\newpage

\begin{sreoutput}
#
# Stage 2: Aligning each sequence to it's best matching model.
#
# output file name            description
# --------------------------  -------------------------------------------
  seed-15.archaea.stk         archaea alignment
  seed-15.archaea.cmalign     archaea cmalign output
  seed-15.bacteria.stk        bacteria alignment
  seed-15.bacteria.cmalign    bacteria cmalign output
  seed-15.eukarya.stk         eukarya alignment
  seed-15.eukarya.cmalign     eukarya cmalign output
  seed-15.scores              list of CM/HMM scores for each sequence
  seed-15.log                 log file (*this* text printed to stdout)
#
# All output files created in directory ./seed-15/
#
# CPU time (search):     00:00:15
# CPU time (alignment):  00:00:21
# CPU time (total):      00:00:37
#
\end{sreoutput}

The newly created alignments are the \prog{.stk} suffixed files. These
were created by \sft{infernal}'s \prog{cmalign} program. The
\prog{cmalign} output is in the \prog{.cmalign} suffixed files.  As in
stage 1, these files were created in the \prog{./seed-15/}
subdirectory. 

\subsubsection{Description of alignments}

Section~\ref{section:output} contains more information
on \sft{ssu-align} output files, but for now we'll focus only on the new
alignments.  Take a look at the archaeal alignment we just created in
\prog{seed-15/seed-15.archaea.stk}.

This alignment includes consensus secondary structure annotation and
is in \emph{Stockholm format}. 
Stockholm format, the native alignment format used by \sft{hmmer} and
\sft{infernal} and the \db{pfam} and \db{rfam}
databases, is documented in detail in the \sft{infernal} user's
guide which is included as a PDF in \prog{infernal-1.01/}

For now, what you need to know about the key features of the alignment file is:
\begin{itemize}

\item The alignment is in an interleaved format, like other
  common alignment file formats such as \sft{clustalw}.
  Lines consist of a name, followed by an aligned sequence;
  the alignment is split into blocks separated by blank lines.

\item Gaps are indicated by the characters ., \_, -, or \verb+~+.
  Many SSU alignments will have large regions of 100\% gaps at the
  beginning and ends of the alignment. 
  This will happen if the sequences are 
  partial SSU sequences, such as those  obtained with PCR
  primers that target well conserved regions within the SSU
  molecule.

\item Special lines starting with {\small\verb+#=GR+} followed by a
  sequence name and then {\small\verb+POST+} contain posterior
  probabilities for each aligned residue for the sequence they
  correspond to. These are confidence estimates in the correctness of
  the alignment.  The POSTX. row indicates the ’tens’ place of the
  confidence estimate while POST.X row indicates the ’ones’ place. So
  the confidence estimate for a residue with 9 in the POSTX. row, and
  7 in the POST.X row to two significant digits is 97\%. This means that
  if you sampled alignments from the posterior distribution of all
  possible alignments of this sequence to the model, about 97% of the
  time that residue would appear in that of the alignment. One special
  case: if the posterior probability is very nearly 100% (it’s
  difficult to be more precise on the exact percentage due to
  numerical precision issues) the annotated posterior values will be
  ``*'' characters in both the tens and ones places. These confidence
  estimates can be used to mask the alignment to remove columns with
  significant fractions of ambiguously aligned residues as demonstrated
  below.

\item A special line starting with {\small\verb+#=GC SS_cons+}
  indicates the secondary structure consensus. Gap characters annotate
  unpaired (single-stranded) columns. Base pairs are indicated by any
  of the following pairs: \verb+<>+, \verb+()+, \verb+[]+, or
  \verb+[]+.

\item A special ``RF'' line starting with {\small\verb+#=GC RF+}
  indicates the consensus, or ReFerence, model. Gaps in the RF line
  are \emph{insert} columns, where at least 1 sequence has at least 1
  inserted residue between two consensus positions. Uppercase residues
  in the RF line are well conserved positions in the model; lowercase
  residues are less well conserved.
\end{itemize}

\begin{comment}
To convert the alignment to fasta format that includes gaps, you can use the
\prog{scripts/stk2aln\_fa.pl} script. 
\end{comment}

\subsection{Creating masks for pruning alignments with the
  \prog{esl-alimanip} program}

If your goal is to use a phylogenetic inference program to build trees
from alignments created by \sft{ssu-align}, you may want to mask
out columns of the alignment that may include misaligned residues
first, and then only run the inference on the remaining columns where
you're confident the alignment is correct. 

As mentioned above, \sft{infernal}'s \prog{cmalign} program
can automatically calculate posterior probabilities that estimate the
level of ambiguity in the alignment of each residue. These can be
interpreted as how confident the model is in each aligned residue in
each sequence and are useful for finding and
removing ambiguously aligned regions prior to phylogenetic inference.
The \prog{esl-alimanip} program can be used to mask
\sft{ssu-align} generated alignments based on these confidence estimates.

I recommend a three step masking process. The first step is to remove
all insert columns from the alignment. Insert columns include residues
that are literally \emph{not aligned}, but rather just inserted in
between the appropriate two consensus positions. 
%(There's more explanation of this in the ``Background'' section).  
Since inserts are
not aligned they should not be included in a phylogenetic inference.
The second step is to create a mask of the consensus model based on
the confidence estimates in the alignment. The final step is to prune
the alignment based on that mask.

Here's an example of using \prog{esl-alimanip} for the three steps. 
First, we'll use \prog{esl-alimanip} to remove all insert columns and
save the new alignment to \prog{seed-15.archaea.noins.stk}:

\user{esl-alimanip -k -o seed-15.archaea.noins.stk seed-15/seed-15.archaea.stk}

%(Note: if \prog{esl-alimanip} is not in your path, you'll have to
%specify the full path to the program. For example:
%\prog{/usr/people/nawrocki/infernal-1.01/easel/miniapps/esl-alimanip}.

The program will create the new alignment file
\prog{seed-15.archaea.noins.stk} almost instantly. 
The \prog{-k} option to \prog{esl-alimanip} tells it to remove all
columns of the alignment that are gaps in the \prog{\#\=GC RF}
annotation. In \sft{ssu-align} generated alignments, gaps in the
\prog{\#\=GC RF} annotation indicate a column is an insert column that
does not correspond to a consensus model position.

Next, we'll generate a mask based on the alignment confidence estimates
in the insert-removed alignment :

\user{esl-alimanip --omask my.mask --p-rf --pfract 0.95 --pthresh
  0.95 \\   -o trash.stk seed-15.archaea.noins.stk}

(Note that the full command includes both lines above. Many of the
commands in the rest of this tutorial are too long for a single line,
and are split up like this one.)
The following will print to the screen:

\begin{sreoutput}
Average posterior value:                            0.98877 (6219 non-gap residues)
Average posterior value in non-gap #=GC RF columns: 0.98877 (6219 non-gap RF residues)

1449 of 1508 RF columns (0.961) pass threshold
\end{sreoutput}

And the mask file \prog{my.mask} is created. Take a look at this
file. It is a single line of text of length $1508$ characters,
containing only \prog{0}s and \prog{1}s. The length is 1508 because
the consensus archaeal model (see section~\ref{section:chap9}) is 1508 residues
long. A \prog{0} at position \prog{n} indicates that consensus position \prog{n}
will excluded (pruned away) when the mask is applied. A \prog{1} at
position \prog{n} indicates that consensus position \prog{n} will be included (not
pruned away) when the mask is applied. In this mask file there are
$1449$ \prog{1}s and $59$ \prog{0s}.

In general the columns where the alignment confidence is high will be
included, and where it is low will be excluded. More specifically, a
column \prog{n} is included if at least 0.95 fraction (95\%)
of the sequences that have a residue (non-gap) in column \prog{n} have a
confidence estimate of 0.95 or better for that residue. All other
columns are excluded.

Using 0.95 for both \prog{--p-fract} and \prog{--p-thresh} are the
recommended options for masking. Section~\ref{section:chap9} 
describes a simple experiment testing how different thresholds affect
alignment accuracy, and using 0.95 for both thresholds resulted in 
high accuracy (see Table~\ref{tbl:kolbe09-pp}).
However, these thresholds can be changed using command-line options to 
\prog{esl-alimanip}. The fraction threshold can be changed to
\prog{<x>} with the \prog{--pfract <x>} and the confidence estimate
threshold can be changed to \prog{<y>} with the \prog{--pthresh <y>}
option.

Note that only non-gap residues are considered when the mask is
generated. This means if a consensus column only has a single residue
in it, and it has high confidence (say 0.98), then that column will be
be included in the mask (not be pruned away). This may be be
undesirable, depending on what type of downstream phylogenetic inference
you are going to do. Many phylogenetic inference programs treat gaps
as missing data, and for such programs you may want to exclude very gappy
columns when masking. 

The \prog{esl-alimanip} program can also take the gappiness of a
column into account when masking. To use the same confidence estimate
thresholds as the previous \prog{esl-alimanip} mask generation
(\prog{--pfract 0.95 --pthresh 0.95}) but \emph{also} exclude any column 
with more than 45\% gaps, execute the command:

\user{esl-alimanip --omask my2.mask -g --gapthresh 0.45 --p-rf
  --pfract 0.95 \\ --pthresh 0.95 -o trash.stk seed-15.archaea.noins.stk}

\begin{sreoutput}
Average posterior value:                            0.98877 (6219 non-gap residues)
Average posterior value in non-gap #=GC RF columns: 0.98901 (6184 non-gap RF residues)

1435 of 1467 RF columns (0.978) pass threshold
\end{sreoutput}

The final line indicates that $1467$ consensus positions had less
than 45\% gaps, of these $1435$ passed the confidence estimate
thresholds. The new mask \prog{my2.mask} is still $1508$ characters
long, but now contains only $1435$ \prog{1}s.

The final step of masking is applying the mask. This is also done with
\prog{esl-alimanip}:

\user{esl-alimanip -k --mask-rf my2.mask -o
  my2.masked.archaea.stk \\ seed-15.archaea.noins.stk}

The new alignment is saved in \prog{my2.masked.archaea.stk}.

The easel program \prog{esl-alistat} can be used to calculate and
display some summary statistics of the new alignment:

\user{esl-alistat my2.masked.archaea.stk}

\begin{sreoutput}
Alignment number:    1
Format:              Stockholm
Number of sequences: 5
Alignment length:    1435
Total # residues:    6069
Smallest:            857
Largest:             1435
Average length:      1213.8
Average identity:    82%
\end{sreoutput}

\subsection{Visualizing alignment masks}

The \prog{esl-ssudraw} program can be used to create a secondary
structure diagram that shows which parts of the SSU consensus model
are included and excluded by the mask. This requires the use of a
postscript template file for the archaeal model. This is provided in 
\\ \prog{ssu-align-0.1/seeds/ss-diagrams/archaea-0p1.ps}. 
%Copy that file into the working directory. To generate a diagram of our
%mask: 
To generate a diagram of our mask:

\user{esl-ssudraw --mask my2.mask --mask-col
  seed-15/seed-15.archaea.stk \\ ../seeds/ss-diagrams/archaea-0p1.ps my2.mask.ps}

The \prog{--mask} option tells \prog{esl-ssudraw} that we are
supplying the mask \prog{my2.mask} for display. The \prog{--mask-col}
option specifies the color scheme for the mask (other types of mask
coloring are demonstrated later in this tutorial. Additionally, 
\prog{esl-ssudraw} takes three command-line arguments. The first is
the alignment file that the mask pertains to. Note that we use
\prog{seed-15.archaea.stk}, this is the original alignment we created,
not the masked alignment. The second argument is the template
postscript file. Here we use \prog{archaea-0p1.ps} because it corresponds
to the \sft{ssu-align} 0.1 archaea CM, which we used to create this
alignment. The third argument is the desired name of the output
postscript file. The \prog{my2.mask.ps} file is created here, which
includes a drawing of the secondary structure diagram. You can either
display this using a postscript viewer, or convert it to a PDF using a
conversion program (I use the \prog{ps2pdf} program that is freely
available for download for most unices). Note that Mac OS/X's
automatic conversion with the \prog{Preview} program is not reliable
for converting these postscript files to PDFs.

The masked secondary structure diagram is displayed on the next page.
The color legend and other information on the diagram is meant to be
self explanatory. \prog{esl-ssudraw} can be used to draw secondary
structures displaying different types of information than masks, as
described next.

\newpage

\begin{figure}
\begin{center}
\includegraphics[width=5.7in]{Figures/my2-mask}
\end{center}
\caption{\textbf{The my2.mask.ps diagram that shows the my2.mask on
    the SSU archaeal consensus structure.}}
\end{figure}

\input{tutorial-draw}

\input{tutorial-trunc}

\input{tutorial-prep}

\input{tutorial-merge}







