\section{Basic tutorial: defining and aligning SSU sequences using \textsc{SSU-align}}

Here is a tutorial walk-through of a small project with
\software{SSUalign}. This tutorial shows how to use the program for
it's most basic and fundamental purpose, creating multiple
alignments of SSU rRNA sequences. 

\subsection{Files used in this tutorial}

The subdirectory \prog{/tutorial-basic} in the \software{SSU-align}
contains the files used in this tutorial, they are:

  \begin{sreitems}{}
  \item[\prog{ssu.default.0p1.cm}] A covariance model (CM) file that
    defines five SSU rRNA CMs: an archael model, a bacterial model, a
    choloroplast model, a eukaryotic model and a metazoan
    mitrochondria model. These are the five default models used by
    \textsc{SSU-align}. These models are explained in section 4.
  \item[\prog{rocks.fa}] SSU rRNA sequences from an environmental
    survey sequencing project of microbes living in the pore space of
    rocks in the Rocky Mountains by J.J. Walker and Norm Pace
    \cite{Walker07}. 
  \item[\prog{1p0.params}] A file containing paths to
    \textsc{infernal} executable files that \textsc{SSU-align} needs
    to run. You will likely need to change these paths to point to
    where you've installed the \prog{cmsearch} and \prog{cmalign}
    programs (these are created in 'infernal-1.0/src/' after building
    \textsc{infernal} version 1.0 with 'sh ./configure; make;') and
    the \prog{esl-sfetch} program (which is created in
    'infernal-1.0/easel/miniapps/' after building \textsc{infernal}
    version 1.0).
  \end{sreitems}

Create a new directory that you can work in, and copy all the files in
\prog{tutorial-basic} there. I'll assume for the following examples that
you've installed the \software{SSU-align} script in your path; if not,
you'll need to give a complete path name to the script
(e.g. something like
\newline
\prog{/usr/people/nawrocki/ssualign/src/ssu-align} 
instead of just \prog{ssu-align}).

The file \prog{rocks.fa} contains 588 SSU sequences
\cite{Walker07}. \textsc{SSU-align} is designed to create structural
alignments of SSU sequences from studies like this one. To run it,
execute the following command:

\user{ssu-align ssu.default.0p1.cm rocks.fa myrun 1p0.params}\\

The program will report on what its doing:

\begin{sreoutput}
#
# Stage 1: Defining SSU start/ends with cmsearch (crude time estimate: 0.3 minutes) ... 
\end{sreoutput}

In stage 1, the program scans the input sequences with each of the
three models in the CM file \prog{abcem.cm}. This has two purposes.
First, it classifies each sequence by determining which model gives
each sequence the highest HMM alignment score. Secondly, it defines
the start and end points of the SSU sequences. 

When this step finishes, you'll see:

\begin{sreoutput}
#
# Stage 1: Defining SSU start/ends with cmsearch (crude time estimate: 1.1 minutes) ...  done.
#
# output file name                                            description
# ----------------------------------------------------------  -----------
  myrun/myrun.tab                                             cmsearch tabular output file with locations/scores of hits using HMM.
  myrun/myrun.archaea.sseq.list                               list of high scoring subseqs to align with archaea CM.
  myrun/myrun.archaea.sseq.fa                                 FASTA file of high scoring subseqs to align with archaea CM (48 sequences).
  myrun/myrun.bacteria.sseq.list                              list of high scoring subseqs to align with bacteria CM.
  myrun/myrun.bacteria.sseq.fa                                FASTA file of high scoring subseqs to align with bacteria CM (341 sequences).
  myrun/myrun.cholorplast.sseq.list                           list of high scoring subseqs to align with cholorplast CM.
  myrun/myrun.cholorplast.sseq.fa                             FASTA file of high scoring subseqs to align with cholorplast CM (199 sequences).
\end{sreoutput}

This lists and briefly describes the 7 new files the script created in
a subdirectory called \prog{myrun/}. The archaeal model
was the best match to $48$ of the $588$
sequences. The \prog{myrun/myrun.archaea.sseq.list} file lists
these sequences. The \prog{myrun/myrun.archaea.sseq.fa}
contains the $48$ subsequences.  $341$ sequences were
scored highest by the bacterial model.  The list
and sequences are in \prog{myrun/myrun.bacteria.sseq.list} and
\prog{myrun/myrun.bacteria.sseq.fa}. The remaining 199 sequences 
were scored highest by the chloroplast model, these are listed in
\prog{myrun/myrun.chloroplast.sseq.list}; the actual sequences are in
\prog{myrun/myrun.chloroplast.sseq.fa}.

There were no sequences that
best-matched the eukaryotic model or the metazoan mitochondria model.

The output of
\textsc{Infernal's} \prog{cmsearch} program is in the file
\prog{myrun/myrun.tab}.

The program will now proceed to stage 2, the alignment stage. This
stage serial progresses through each model that had at least 1
matching sequence and aligns the sequences to the model using both
structure and sequence conservation. A time estimate is provided for
each stage.

\begin{sreoutput}
#
# Stage 2. Aligning sequences.
#
# stage  cm                   seq file                                                         nseq  est min
# -----  -------------------  ----------------------------------------------------------    -------  -------
   1/ 5  archaea              0615-1/0615-1.archaea.sseq.fa                                      48     0.81
   2/ 5  bacteria             0615-1/0615-1.bacteria.sseq.fa                                    341     6.32
   3/ 5  cholorplast          0615-1/0615-1.cholorplast.sseq.fa                                 199     3.38
   4/ 5  eukarya              NONE (no matching seqs)                                             0     0.00
   5/ 5  mitochondria-animal  NONE (no matching seqs)                                             0     0.00
\end{sreoutput}

After the alignment stage ends there will be three new alignment files:
\prog{myrun/myrun.m1.cmalign.stk} and \prog{myrun/myrun.m2.cmalign.stk}. 

Take a look at the archaeal alignment in
\prog{myrun/myrun.m1.cmalign.stk}. 

This alignment includes consensus secondary structure annotation and
is in \emph{Stockholm format}. 
Stockholm format, the native alignment format used by \software{hmmer} and
\software{Infernal} and the \database{Pfam} and \database{Rfam}
databases, is documented in detail in the \software{Infernal} User's
Guide which is included in this distribution in
\prog{infernal/documentation/userguide.pdf}.

For now, what you need to know about the key features of the input file is:
\begin{itemize}
\item The alignment is in an interleaved format, like other
common alignment file formats such as \software{clustalw}.
Lines consist of a name, followed by an aligned sequence;
long alignments are split into blocks separated by blank lines.
\item Each sequence must have a unique name that has zero spaces in it. (This is important!)
\item For residues, any one-letter IUPAC nucleotide code is accepted,
      including ambiguous nucleotides. Case is ignored; residues
      may be either upper or lower case.
\item Gaps are indicated by the characters ., \_, -, or \verb+~+.
      (Blank space is not allowed.)
\item A special line starting with {\small\verb+#=GC SS_cons+} indicates
      the secondary structure consensus. Gap characters annotate
      unpaired (single-stranded) columns. Base pairs are indicated
      by any of the following pairs: \verb+<>+, \verb+()+, \verb+[]+,
      or \verb+[]+. No pseudoknots are allowed; the
      open/close-brackets notation is only unambiguous for strictly
      nested base-pairing interactions.
\item The file begins with the special tag line
      {\small\verb+# STOCKHOLM 1.0+}, and ends with {\small\verb+//+}.
\end{itemize}

To convert the alignment to fasta format that includes gaps, you can use the
\prog{scripts/stk2aln\_fa.pl} script. 

NOT SURE WHAT TO WRITE ABOUT THE ALIGNMENT!

\subsection{Pruning the alignment based on probabilistic confidence
  estimates}
