\section{Extended Tutorial}

This section covers some additional features of \textsc{ssu-align}
that may be useful to some users but that are not covered in the basic
tutorial section. 

\subsection{Creating a truncated model of a specific region of SSU rRNA}

Many SSU rRNA sequencing studies target a specific region of the SSU
rRNA molecule using PCR primers at the boundaries of that region. For
such studies it is recommended to build a new CM that only models the
region of the molecule targetted by the study. There are two reasons
for this. The first is speed; the running time of \textsc{ssu-align}
decreases as the model size it's using decreases. The second reason is
that aligning a SSU subsequence to a model of only the region that
subsequence is derived from, relative to a model of the entire SSU
molecule, should slightly increase alignment accuracy. This is because
the uncertainty of what region of the full molecule the subsequence
should align to is eliminated. In this section I'll demonstrate how to
create a CM of a specific region of SSU and use it to create
alignments. 

For this example imagine our study is only targetting bacterial SSU
rRNA. We will use the bacterial SSU seed alignment that is included
with \textsc{ssu-align} as a starting point for creating our new,
truncated CM. The first step is to determine what consensus positions
in the bacterial seed alignment's consensus structure the targetted
region corresponds to. The consensus structure of the bacterial seed
alignment is shown in the ``Models'' section on page X.

Create a new directory and copy the bacterial seed alignment in
\prog{seeds/bacteria-0p1.stk}, the default parameters
file \prog{sa-0p1.params}, and the file \prog{tutorial/partial.fa}
there.

Let's say our 5' primer begins at consensus position 35 and our 3'
primer ends at position 397.  In practice, you'll have to manually
find your primer site and determine their positions on this consensus
structure. On the structure diagrams in the ``Models'' section, every
hundredth residue is numbered, and every tenth residue is marked with
a tick mark, which should help you find the relevant positions.  If
you have a subsequence that exactly spans from one primer to another,
you can align it to the appropriate \textsc{ssu-align} model, then
number the consensus positions of that alignment with
\prog{esl-alimanip --num-rf} and examine the numbered alignment to
determine the primer positions.

\begin{srefaq}{If I'm creating a truncated model for sequences derived
    using specific primers, should I include the primer sequences
    within the new model or not?} You should include the primers
  because they will help the program correctly align each
  sequence. The primer sites are very highly conserved so they are
  simple for the program to correctly align and anchor the alignment
  of the rest of the sequence.
\end{srefaq}

For this example, the first step towards creating a truncated model is
to create a truncated seed alignment that only models between
consensus positions 35 and 397. We can do this with the
\prog{esl-alimanip} program using the full bacterial seed alignment as
input:

\user{esl-alimanip --start-rf 35 --end-rf 397 -o bac-35-397.stk bacteria-0p1.stk}

This command creates a new alignment called \prog{back-35-397.stk} which
includes the subset of the columns from \prog{bacteria-0p1.stk} that lie
between consensus positions 35 and 397 inclusively.

The next step is to build a new model from this new alignment with
\textsc{infernal} 1.0's \prog{cmbuild} program. If this program is in
your path, you can execute it with \prog{cmbuild}, otherwise you'll
need to provide the full path. We will specify the name we want
to give the model with the \prog{-n} flag:

\user{cmbuild -n bac-35-397 --enone --gapthresh 0.8 bac-35-397.cm bac-35-397.stk}

\begin{srefaq}{Why did you specify \prog{--enone} and
    \prog{--gapthresh 0.8} as command-line flags to \prog{cmbuild}?}
    These are the recommended ``best-practice'' options for building
    models for SSU alignment. The \prog{--enone} flag tells the
    program to turn off entropy-weighting, a parameterization
    technique used to make CM homology search more sensitive
    \cite{Nawrocki07} but that seems slightly detrimental to CM
    alignment accuracy with SSU rRNA models. The \prog{--gapthresh
      0.8} flag tells the program to define any column that has less
    than 80\% gaps in the seed as a consensus column. Different values
    than 0.8 could be used here, but 0.8 empirically seems to yield
    good performance for SSU alignment. The \prog{--enone} and
    \prog{--gapthresh 0.8} flags were used to build
    \textsc{ssu-align}'s five default models in the \prog{seeds/}
    subdirectory of the package.
\end{srefaq}

Now you can begin using your new model \prog{bac-35-397.cm} to align
SSU sequences. You have two options.  You can either use your new
model by itself as the only model in an \prog{ssu-align} run, or you
can combine it with other models to create a multi-model file to use
with \prog{ssu-align}. The former option is recommended if you expect
all of your sequences to match the truncated model, e.g. in this case,
be bacterial SSU subsequences that map close to the 35-397
region. I'll run through an example of this below. The latter option,
combining this model with others, is recommended if only a subset of
the sequences you will analyze are expected to match the truncated
model. In that case, the other models you combine the truncated model
should span the diversity of the other sequences you expect in your
sequence dataset. (An example of using a multi-model file with
\textsc{ssu-align} is demonstrated in the basic tutorial section).

Imagine we expect all the sequences in our sequence dataset are
bacterial sequences that match near the 35-397 region. An example
sequence file with 8 such sequences derived from larger sequences in
the \prog{rocks.fa} file used in the basic tutorial is included in
\prog{partial.fa}. I artificially added 10 random residues to the 5\' and 3\'
ends of the 8th sequence to demonstrate that the program can 
trim residues it deems nonhomologous to the model from the ends of
the sequences before alignment.

\user{ssu-align bac-35-397.cm partial.fa single sa-0p1.params}

The program takes about 2 seconds to run. 

Take a look at the \prog{single.scores} file in the \prog{single/}
subdirectory:

\begin{sreoutput}
#                                                      best-matching model                 
#                                       -------------------------------------------------  
#     idx  sequence name                model name   beg   end    CM sc   struct   HMM sc
# -------  ---------------------------  ----------  ----  ----  -------  -------  -------
        1  gi|146141790|gb|EF522294.1|  bac-35-397     1   344   403.20   100.11   312.76
        2  gi|146141797|gb|EF522301.1|  bac-35-397     1   348   463.37    69.69   402.70
        3  gi|146141805|gb|EF522309.1|  bac-35-397     1   315   475.02    63.31   416.05
        4  gi|146141812|gb|EF522316.1|  bac-35-397     1   339   480.57    68.35   423.81
        5  gi|146141828|gb|EF522332.1|  bac-35-397     1   336   450.46    70.82   392.66
        6  gi|146141831|gb|EF522335.1|  bac-35-397     1   332   380.82    94.56   297.11
        7  gi|146141832|gb|EF522336.1|  bac-35-397     1   334   476.07    69.57   415.65
        8  gi|146141837|gb|EF522341.1|  bac-35-397    11   363   338.11    83.56   285.87
\end{sreoutput}

Note how the alignment of the final sequence begins at position 11 and
ends at 363, truncated the first and last 10 nonhomologous residues
that I had manually added (that sequence is 373 residues in
\prog{partial.fa}).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Creating secondary structure diagrams that
  display alignment statistics}

SSU rRNA alignments are large and difficult to view in a meaningful
way. \textsc{ssu-align} includes a program \prog{esl-ssudraw} for
creating secondary structure diagrams that display statistics of a
particular alignment on the consensus secondary structure used to
align the model. To use \prog{esl-ssudraw} requires a template
postscript file of the consensus secondary structure. The template
files for the 5 default \textsc{ssu-align} version 0.1 seed models are
included, but constructing these required a significant amount of
work, and creating your own templates for different models would be
difficult. 

Let's create a diagram that shows the information content of each
position of the archaeal alignment created in the basic tutorial of
sequences from from the Walker and Pace
rocks study:

TO DO!

\prog{esl-ssudraw rocks/rocks.archaea.cmalign.stk ../../seeds/ss-diagrams/archaea-0p1.ps my.ps}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Creating new SSU models that cover
  different ``tree space'' than the default models}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Splitting up large alignment jobs}

If you have access to a compute cluster you can partition your input
dataset and run \textsc{ssu-align} in parallel on multiple machines. A
motivated user could certainly write their own scripts to do this, but
the \prog{ssu-align} program has options to facilitate this
parallelization so you don't have to.  It also will create a script
that will merge the separate alignments created for each partition
into a single master alignment.

In this section we'll walk through an example of how to do this for a
small dataset.  The sequence file \prog{tutorial/seed-50.fa} includes
50 randomly chosen sequences from the complete set of seed sequences
from the 5 default models of \textsc{ssu-align} (there are 282 total
seed sequences, as shown in the table of seed alignment statistics in
section 3). Create a new directory within \prog{tutorial/} and copy
\prog{tutorial/seed-50.fa}, the file
\prog{tutorial/janelia-sa-0p1.params} and the default parameters file
\prog{sa-0p1.params} there.

There are two ways \textsc{ssu-align} can split up a job into multiple
jobs to run in parallel. You can either specify the number of jobs
(\prog{<x>}) to create, or the number of sequences (\prog{<y>}) you
want each job to handle. These are invoked using the command-line
options \prog{-c <x>} and \prog{-n <y>} respectively.  In this example
we'll use the \prog{-c} method.

One important modification should be made to the parameters file
before you use \textsc{ssu-align} to split up a large job to run in
parallel on a cluster. You'll want to create a new parameters file
(you can use the default \prog{sa-0p1.params} as a starting point) and
add two lines to it defining a prefix string and suffix string that
should appear before and after the \prog{ssu-align} call if you were
submitting it as a job for the cluster to your local job
scheduler. For example, look at the final 3 lines of the file
\prog{tutorial/janelia-sa-0p1.params}:

\begin{sreoutput}
\$cluster\_prefix = "qsub -N ssualign -o ssualign.out -b y -cwd -V -j y '";
\$cluster\_suffix = "'";
1;
\end{sreoutput}

The definitions of the variables \prog{\$cluster\_prefix} and
\prog{\$cluster\_suffix} tell \prog{ssu-align} what prefix and suffix
strings, respectively, to add to the \prog{ssu-align} runs it will
create for each partition. These specific strings correspond to the
format used by the SGE (Sun Grid Engine) \prog{qsub} program we use
for scheduling jobs on our cluster here at Janelia Farm in Virginia.
Importantly, you'll need to change these to the format required by
your own compute resources.

Because our example dataset here is only 50 sequences we really don't
need a cluster to run it, so we don't need to define the
\prog{\$cluster\_prefix} and \prog{\$cluster\_suffix} variables in our
parameters file. We'll use the default \prog{sa-0p1.params}.

Let's say we want to split up our 50 sequences into 5 separate files
of 10 sequences each and run each set of 10 independently. 
To execute the script:

\user{ssu-align -c 5 ../../seeds/ssu5-0p1.cm seed-50.fa seed-50-c5 ../../sa-0p1.params}

The program finishes in about 1 second. It will print information on
the files it has created:

\begin{comment}
# ssu-align :: define and align SSU rRNA sequences
# SSU-ALIGN 0.1 (June 2009)
# Copyright (C) 2009 HHMI Janelia Farm Research Campus
# Freely distributed under the GNU General Public License (GPLv3)
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# command: /groups/eddy/home/nawrockie/ssualign/ssu-align -F -c 5 ../../seeds/ssu5-0p1.cm seed-50.fa seed-50-c5 ../../sa-0p1.params
# date:    Thu Jun 18 16:27:21 2009
#
\end{comment}
\begin{sreoutput}
# Prep mode: Splitting up ssu-align job into 5 smaller jobs.
#
# output file name     description                                                 
# -------------------  ------------------------------------------------------------
  seed-50.fa.1         partition 1 fasta sequence file
  seed-50.fa.2         partition 2 fasta sequence file
  seed-50.fa.3         partition 3 fasta sequence file
  seed-50.fa.4         partition 4 fasta sequence file
  seed-50.fa.5         partition 5 fasta sequence file
  seed-50-c5.sh        shell script that will run ssu-align 5 times
  seed-50-c5.merge.pl  perl script to merge alignments when seed-50-c5.sh completes
  seed-50-c5.log       log file (*this* text printed to stdout)
#
# All output files created in directory ./seed-50-c5/
#
\end{sreoutput}

The first five fasta files are partitions of the original sequence
file \prog{seed-50.fa}. Each has 10 sequences in it. The file
\prog{seed-50-c5.sh} is a shell script that will execute
\prog{ssu-align} five times, once for each of the partitions. 
If we had defined \prog{\$cluster\_prefix} and
\prog{\$cluster\_suffix} this script would include those strings to
allow submission to a cluster. Since we did not our script will simply
execute \prog{ssu-align} five times in succession. The file
\prog{seed-50-c5.merge.pl} is a perl script we can use to merge the
individual alignments created by each of the five jobs together.
The \prog{seed-50-c5.log} file contains the exact text that was just
printed to the screen. All of these files were created in a new
directory called \prog{seed-50-c5}.

The next step is to run the \prog{seed-50-c5.sh} script, which will
execute \prog{ssu-align} five times. First, enter the
\prog{seed-50-c5/} directory, then type:

\prog{sh seed-50-c5.sh}

The output of the five \prog{ssu-align} runs will begin to print to
the screen. When one run finishes, the next will immediately
start. Each run should take about 30 seconds, so all five take a few
minutes. Once they're done we can move to the next step: merging the
alignments. If we had defined \prog{\$cluster\_prefix} and
\prog{\$cluster\_suffix}, each line of our script would have submitted
a separate job to our cluster. The script would finish quickly, but
the jobs would still be running. \textbf{\emph{Important: }} you'll
need to wait until \emph{all} the jobs are finished running on the
cluster before attempting to merge the alignments. 

Continuing with our example, to merge the alignments:
\prog{perl seed-50-c5.merge.pl}

This script will create a merged alignment for each model in the
original cm file (\prog{ssu5-0p1.cm} in this case) that was the
best-matching model for at least 1 sequence in any of the five runs.
For each such model, the script will merge two alignments at a time
using \textsc{infernal}'s \prog{cmalign} program (see the next section
on ``Merging multiple alignments together'' for more information).
done) until all a single alignment with all the sequences is
created. 

For example, in this case the alignment from jobs 1 and 2 would be
merged first, then from jobs 3 and 4. Now three alignments exist, one
from jobs 1+2, one from 3+4 and one from 5. Next 1+2 are merged with
3+4, and finally the resulting alignment of 1+2+3+4 is merged with the
alignment from job 5. If you look at the code in
\prog{seed-50-c5.merge.pl} it may look complex, but this simple binary
merging is all that it is doing. The intermediate alignments are
deleted once they are no longer needed as the script proceeds. If you
have a large number of alignments to merge and/or a large number of total
sequences this script may take a long time. For more information on
running times see the ``Performance'' section.

In this case, 5 merged alignments are created, as reported by the
script:

\begin{sreoutput}
Merged alignment for archaea     CM saved to seed-50-c5.1-5.archaea.merged.stk
Merged alignment for bacteria    CM saved to seed-50-c5.1-5.bacteria.merged.stk
Merged alignment for chloroplast CM saved to seed-50-c5.1-5.chloroplast.merged.stk
Merged alignment for eukarya     CM saved to seed-50-c5.1-5.eukarya.merged.stk
Merged alignment for metamito    CM saved to seed-50-c5.1-5.metamito.merged.stk
\end{sreoutput}

These alignments will be \%100 identical to the alignments that would
have been created if we had not split up this job, but rather aligned
all 50 sequences with a single run of \prog{ssu-align}.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Merging multiple alignments together}

The \prog{cmalign} program of \textsc{infernal} is capable of merging
two alignments into one. The two alignments must have both been created by
\prog{cmalign} (and the same version of \prog{cmalign}, 1.0 or later),
and must have been created using the same exact CM. This ability is
potentially useful for saving time when aligning a large number of
sequences if you have access to a compute cluster, as described in the
previous section (``Splitting up large alignment jobs''), or if you
want to merge an existing reference alignment with a newly created
one. Combining a new alignment with a reference may be useful in
downstream phylogenetic analysis if you know the classification of the
sequences in the reference alignment.

Imagine we wanted to merge the archaeal sequences from the
\prog{rocks.fa} sequence dataset from the basic tutorial with 
an alignment of the archaeal seed sequences. 

Merging alignments only makes sense and saves time if you've already
computed the two alignments you want to merge. For this example I've
provided the two alignments in the \prog{/tutorial} subdirectory:

\begin{description}
\item[\emprog{rocks.archaea.stk}]
  An alignment of the archaeal sequences from the \prog{rocks.fa}
  dataset. The basic tutorial steps through how to create this file.

\item[\emprog{seed.archaea.stk}]
  An alignment of the archaeal seed sequences, created with the
  command \prog{ssu-align ../seeds/archaea-0p1.cm ../seeds/archaea-0p1.fa seed ../ssualign.0p1.params}
  (Note: this file is NOT identical to the seed alignment
  \prog{archaea-0p1.stk}. I've realigned those sequences to the
  archaeal model to obtain posterior probability confidence estimates
  in the alignment. 
\end{description}

If \prog{cmalign} version 1.0 is in your path, you 
can merge these two alignments into a single alignment called
\prog{seed-rocks.archaea.stk} with

\user{cmalign --merge -o seed-rocks.archaea.stk ../seeds/archaea-0p1.cm seed.archaea.stk rocks.archaea.stk}

The resulting alignment will be 100\% identical to the alignment
\prog{cmalign} would have created if it were used to align a single
sequence file with the seed sequences and rocks sequences together.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{? Manipulating alignments with the
  \prog{esl-alimanip} program ?}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

