\section{Extended Tutorial}

This section covers some additional features of \textsc{ssu-align}
that may be useful to some users but that are not covered in the basic
tutorial section. 

\begin{comment}
\subsection{Files used in this tutorial}

The subdirectory \prog{/tutorial-extended} in the \software{ssu-align}
distribution contains the files used in the tutorial:

  \begin{sreitems}{}
  \item[\prog{ssu5-0p1.cm}] A covariance model (CM) file that
    defines five SSU rRNA CMs: an archael model, a bacterial model, a
    choloroplast wmodel, a eukaryotic model and a metazoan
    mitrochondria model. These are the five default models used by
    textsc{SSUalign}.
  \item[\prog{rocks.fa}] SSU rRNA sequences from the an environmental
    survey sequencing project of microbes living in the pore space of
    rocks in the Rocky Mountains by J.J. Walker and Norm Pace
    \cite{Walker07}. 
  \end{sreitems}

Create a new directory that you can work in, and copy all the files in
\prog{tutorial-extended} there. I'll assume for the following examples
that you've installed the \software{ssu-align} programs in your path;
if not, you'll need to give a complete path name to the programs
(e.g. something like
\newline
\prog{/usr/people/nawrocki/ssualign/src/ssu-align} 
instead of just \prog{ssu-align}).
\end{comment}

\subsection{Creating a truncated model of a specific region of SSU rRNA}

Many SSU rRNA sequencing studies target a specific region of the SSU
rRNA molecule using PCR primers at the boundaries of that region. For
such studies it is recommended to build a new CM that only models the
region of the molecule targetted by the study. There are two reasons
for this. The first is speed; the running time of \textsc{ssu-align}
decreases as the model size it's using decreases. The second reason is
that aligning a SSU subsequence to a model of only the region that
subsequence is derived from, relative to a model of the entire SSU
molecule, should slightly increase alignment accuracy. This is because
the uncertainty of what region of the full molecule the subsequence
should align to is eliminated. In this section I'll demonstrate how to
create a CM of a specific region of SSU and use it to create
alignments.

For this example imagine our study is only targetting bacterial SSU
rRNA. We will use the bacterial SSU seed alignment that is included
with \textsc{ssu-align} as a starting point for creating our new,
truncated CM. The first step is to determine what consensus positions
in the bacterial seed alignment's consensus structure the targetted
region corresponds to. The consensus structure of the bacterial seed
alignment is shown in the ``Models'' section on page X.

Imagine for this example our 5' primer begins at consensus position 35
and our 3' primer ends at position 397.  In practice, you'll have to
manually find your primer site and determine their positions on this
consensus structure. On the structure diagrams in the ``Models''
section, every hundredth residue is numbered, and every tenth residue
is marked with a tick mark, which should help you find the relevant
positions.  If you have a subsequence that exactly spans from one
primer to another, you can align it to the appropriate
\textsc{ssu-align} model, then number the consensus positions of that
alignment with \prog{esl-alimanip --num-rf} and examine the numbered
alignment to determine the primer positions.

\begin{srefaq}{If I'm creating a truncated model for sequences derived
    using specific primers, should I include the primer sequences
    within the new model or not?} You should include the primers
  because they will help the program correctly align each
  sequence. The primer sites are very highly conserved so they are
  simple for the program to correctly align and anchor the alignment
  of the rest of the sequence.
\end{srefaq}

For this example, the first step towards creating a truncated model is
to create a truncated seed alignment that only models between
consensus positions 35 and 397. We can do this with the
\prog{esl-alimanip} program using the full bacterial seed alignment as
input:

\user{esl-alimanip --start-rf 35 --end-rf 397 -o bac-35-397.stk bacteria-0p1.stk}

This command creates a new alignment called \prog{mybac.stk} which
includes the subset of the columns from \prog{bacteria-0p1.stk} that lie
between consensus positions 35 and 397 inclusively.

The next step is to build a new model from this new alignment with
\textsc{infernal} 1.0's \prog{cmbuild} program. If this program is in
your path, you can execute it with \prog{cmbuild}, otherwise you'll
need to provide the full path. We can also specify the name:

\user{cmbuild -n bac-35-397 --enone --gapthresh 0.8 bac-35-397.cm bac-35-397.stk}

\begin{srefaq}{Why did you specify \prog{--enone} and
    \prog{--gapthresh 0.8} as command-line flags to \prog{cmbuild}?}
    These are the recommended ``best-practice'' options for building
    models for SSU alignment. The \prog{--enone} flag tells the
    program to turn off entropy-weighting, a parameterization
    technique used to make CM homology search more sensitive
    \cite{Nawrocki07} but that seems slightly detrimental to CM
    alignment accuracy with SSU rRNA models. The \prog{--gapthresh
      0.8} flag tells the program to define any column that has less
    than 80\% gaps in the seed as a consensus column. Different values
    than 0.8 could be used here, but 0.8 empirically seems to yield
    good performance for SSU alignment. The \prog{--enone} and
    \prog{--gapthresh 0.8} flags were used to build
    \textsc{ssu-align}'s five default models in the \prog{seeds/}
    subdirectory of the package.
\end{srefaq}

Now you can begin using your new model \prog{bac-35-397.cm} to align
SSU sequences. You have two options.  You can either use your new
model by itself as the only model in an \prog{ssu-align} run, or you
can combine it with other models to create a multi-model file to use
with \prog{ssu-align}. The former option is recommended if you expect
all of your sequences to match the truncated model, e.g. in this case,
be bacterial SSU subsequences that map close to the 35-397
region. I'll run through an example of this below. The latter option,
combining this model with others, is recommended if only a subset of
the sequences you will analyze are expected to match the truncated
model. In that case, the other models you combine the truncated model
should span the diversity of the other sequences you expect in your
sequence dataset. (An example of using a multi-model file with
\textsc{ssu-align} is demonstrated in the basic tutorial section).

Imagine we expect all the sequences in our sequence dataset are
bacterial sequences that match near the 35-397 region. An example
sequence file with 8 such sequences derived from larger sequences in
the \prog{rocks.fa} file used in the basic tutorial is included in
\prog{partial.fa}. I artificially added 10 random residues to the 5\' and 3\'
ends of the 8th sequence to demonstrate that the program can 
trim residues it deems nonhomologous to the model from the ends of
the sequences before alignment.

\user{ssu-align bac-35-397.cm partial.fa single ssualign.0p1.params}

The program takes about 2 seconds to run. 

Take a look at the \prog{single.scores} file in the \prog{single/}
subdir:

\begin{sreoutput}
#                                                      best-matching model                 
#                                       -------------------------------------------------  
#     idx  sequence name                model name   beg   end    CM sc   struct   HMM sc
# -------  ---------------------------  ----------  ----  ----  -------  -------  -------
        1  gi|146141790|gb|EF522294.1|  bac-35-397     1   344   403.20   100.11   312.76
        2  gi|146141797|gb|EF522301.1|  bac-35-397     1   348   463.37    69.69   402.70
        3  gi|146141805|gb|EF522309.1|  bac-35-397     1   315   475.02    63.31   416.05
        4  gi|146141812|gb|EF522316.1|  bac-35-397     1   339   480.57    68.35   423.81
        5  gi|146141828|gb|EF522332.1|  bac-35-397     1   336   450.46    70.82   392.66
        6  gi|146141831|gb|EF522335.1|  bac-35-397     1   332   380.82    94.56   297.11
        7  gi|146141832|gb|EF522336.1|  bac-35-397     1   334   476.07    69.57   415.65
        8  gi|146141837|gb|EF522341.1|  bac-35-397    11   363   338.11    83.56   285.87
\end{sreoutput}

Note how the alignment of the final sequence begins at position 11 and
ends at 363, truncated the first and last 10 nonhomologous residues
that I had manually added (that sequence is 373 residues in
\prog{partial.fa}).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Advanced tutorial: creating secondary structure diagrams that
  display alignment statistics}

SSU rRNA alignments are large and difficult to view in a meaningful
way. \textsc{ssu-align} includes a program \prog{esl-ssudraw} for
creating secondary structure diagrams that display statistics of a
particular alignment on the consensus secondary structure used to
align the model. To use \prog{esl-ssudraw} requires a template
postscript file of the consensus secondary structure. The template
files for the 5 default \textsc{ssu-align} version 0.1 seed models are
included, but constructing these required a significant amount of
work, and creating your own templates for different models would be
difficult. 

Let's create a diagram that shows the information content of each
position of the archaeal alignment created in the basic tutorial of
sequences from from the Walker and Pace
rocks study:

TO DO!

\prog{esl-ssudraw rocks/rocks.archaea.cmalign.stk ../../seeds/ss-diagrams/archaea-0p1.ps my.ps}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Advanced tutorial: creating new SSU models that cover
  different ``tree space'' than the default models}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Advanced tutorial: splitting up large alignment jobs}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Advanced tutorial: merging multiple alignments together}

The \prog{cmalign} program of \textsc{infernal} is capable of merging
two alignments into one. The two alignments must have both been created by
\prog{cmalign} (and the same version of \prog{cmalign}, 1.0 or later),
and must have been created using the same exact CM. This ability is
potentially useful for saving time when aligning a large number of
sequences if you have access to a compute cluster, as described in the
previous section (``Splitting up large alignment jobs''), or if you
want to merge an existing reference alignment with a newly created
one. Combining a new alignment with a reference may be useful in
downstream phylogenetic analysis if you know the classification of the
sequences in the reference alignment.

Imagine we wanted to merge the archaeal sequences from the
\prog{rocks.fa} sequence dataset from the basic tutorial with 
an alignment of the archaeal seed sequences. 

Merging alignments only makes sense and saves time if you've already
computed the two alignments you want to merge. For this example I've
provided the two alignments in the \prog{/tutorials} subdirectory:

\begin{description}
\item[\emprog{rocks.archaea.stk}]
  An alignment of the archaeal sequences from the \prog{rocks.fa}
  dataset. The basic tutorial steps through how to create this file.

\item[\emprog{seed.archaea.stk}]
  An alignment of the archaeal seed sequences, created with the
  command \prog{ssu-align ../seeds/archaea-0p1.cm ../seeds/archaea-0p1.fa seed ../ssualign.0p1.params}
  (Note: this file is NOT identical to the seed alignment
  \prog{archaea-0p1.stk}. I've realigned those sequences to the
  archaeal model to obtain posterior probability confidence estimates
  in the alignment. 
\end{description}

If \prog{cmalign} version 1.0 is in your path, you 
can merge these two alignments into a single alignment called
\prog{seed-rocks.archaea.stk} with

\user{cmalign --merge -o seed-rocks.archaea.stk ../seeds/archaea-0p1.cm seed.archaea.stk rocks.archaea.stk}

The resulting alignment will be 100\% identical to the alignment
\prog{cmalign} would have created if it were used to align a single
sequence file with the seed sequences and rocks sequences together.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{? Advanced tutorial: manipulating alignments with the
  \prog{esl-alimanip} program ?}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{? Benchmarking \textsc{ssu-align} alignment accuracy ?}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

